<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Fast, Epiallele-Aware Methylation Reporter • epialleleR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Fast, Epiallele-Aware Methylation Reporter">
<meta property="og:description" content="Epialleles are specific DNA methylation patterns that are
  mitotically and/or meiotically inherited. This package calls hypermethylated
  epiallele frequencies at the level of genomic regions or individual cytosines
  in next-generation sequencing data using binary alignment map (BAM) files as
  an input. Other functionality includes extracting methylation patterns,
  computing the empirical cumulative distribution function for per-read beta
  values, and testing the significance of the association between epiallele
  methylation status and base frequencies at particular genomic positions
  (SNPs).">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">epialleleR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.9.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="articles/epialleleR.html">Get started</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/BBCG/epialleleR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="fast-epiallele-aware-methylation-reporter-">Fast, epiallele-aware methylation reporter <img align="right" src="inst/epialleleR_logo.svg"><a class="anchor" aria-label="anchor" href="#fast-epiallele-aware-methylation-reporter-"></a>
</h1></div>
<p><a href="https://github.com/BBCG/epialleleR/actions" class="external-link"><img src="https://github.com/BBCG/epialleleR/workflows/R-CMD-check-bioc/badge.svg"></a> <a href="https://codecov.io/gh/BBCG/epialleleR" class="external-link"><img src="https://codecov.io/gh/BBCG/epialleleR/branch/devel/graph/badge.svg"></a> <a href="https://bioconductor.org/packages/release/bioc/html/epialleleR.html" class="external-link"><img src="https://bioconductor.org/shields/years-in-bioc/epialleleR.svg"></a></p>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>DISCLAIMER: This is a work in progress, however the package is already usable, and the obtained experimental results have been published. Main methods (<em><code>preprocessBam</code></em>, <em><code>generateCytosineReport</code></em>, <em><code>generateBedReport</code></em>) won’t change. The <em><code>generateVcfReport</code></em> method might at some point be improved to include variable-length sequence variations, while <em><code>generateBedEcdf</code></em> should be considered somewhat experimental and may undergo significant changes or be substituted with some other method in the future.</p>
<p><em><code>epialleleR</code></em> is an R package for calling hypermethylated variant epiallele frequencies (VEF) at the level of genomic regions or individual cytosines in next-generation sequencing data using binary alignment map (BAM) files as an input. Other functionality includes extracting methylation patterns, computing the empirical cumulative distribution function for per-read beta values, and testing the significance of the association between epiallele methylation status and base frequencies at particular genomic positions (SNPs).</p>
<div class="section level3">
<h3 id="current-features">Current Features<a class="anchor" aria-label="anchor" href="#current-features"></a>
</h3>
<ul>
<li>conventional reporting of cytosine methylation (<em><code>generateCytosineReport</code></em>)</li>
<li>calling the hypermethylated variant epiallele frequency (VEF) at the level of genomic regions (<em><code>generate[Bed|Amplicon|Capture]Report</code></em>) or individual cytosines (<em><code>generateCytosineReport</code></em>)</li>
<li>extracting methylation patterns for genomic region of interest (<em><code>extractPatterns</code></em>)</li>
<li>assessing the distribution of per-read beta values for genomic regions of interest (<em><code>generateBedEcdf</code></em>)</li>
<li>testing for the association between epiallele methylation status and sequence variations (<em><code>generateVcfReport</code></em>)</li>
</ul>
</div>
<div class="section level3">
<h3 id="recent-improvements">Recent improvements<a class="anchor" aria-label="anchor" href="#recent-improvements"></a>
</h3>
<div class="section level5">
<h5 id="v110">v1.10<a class="anchor" aria-label="anchor" href="#v110"></a>
</h5>
<ul>
<li>inputs both single-end and paired-end sequencing alignments</li>
</ul>
</div>
<div class="section level5">
<h5 id="v14">v1.4<a class="anchor" aria-label="anchor" href="#v14"></a>
</h5>
<ul>
<li>significant speed-up</li>
<li>method to visualize methylation patterns</li>
</ul>
</div>
<div class="section level5">
<h5 id="v12">v1.2<a class="anchor" aria-label="anchor" href="#v12"></a>
</h5>
<ul>
<li>even faster and more memory-efficient BAM loading (by means of HTSlib)</li>
<li>min.baseq parameter to reduce the effect of low quality bases on methylation or SNV calling (in v1.0 the output of <em><code>generateVcfReport</code></em> was equivalent to the one of <code>samtools mpileup -Q 0 ...</code>)</li>
</ul>
<p>check out NEWS for more!</p>
<hr>
</div>
</div>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="section level3">
<h3 id="install-via-bioconductor">install via Bioconductor<a class="anchor" aria-label="anchor" href="#install-via-bioconductor"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"epialleleR"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="install-the-latest-version-via-install_github">Install the latest version via install_github<a class="anchor" aria-label="anchor" href="#install-the-latest-version-via-install_github"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/" class="external-link">devtools</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"BBCG/epialleleR"</span>, build_vignettes<span class="op">=</span><span class="cn">FALSE</span>,</span>
<span>  repos<span class="op">=</span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/repositories.html" class="external-link">repositories</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  dependencies<span class="op">=</span><span class="cn">TRUE</span>, type<span class="op">=</span><span class="st">"source"</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="using-the-package">Using the package<a class="anchor" aria-label="anchor" href="#using-the-package"></a>
</h2>
<p>Please read <em><code>epialleleR</code></em> vignettes at <a href="https://bbcg.github.io/epialleleR/articles/epialleleR.html" class="external-link">GitHub pages</a> or within the R environment: <code><a href="articles/epialleleR.html">vignette("epialleleR", package="epialleleR")</a></code>, or consult the function’s help pages for the extensive information on usage, parameters and output values.</p>
<p>Very brief synopsis:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/BBCG/epialleleR" class="external-link">epialleleR</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># external files</span></span>
<span><span class="va">amplicon.bam</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"amplicon010meth.bam"</span>,</span>
<span>                            package<span class="op">=</span><span class="st">"epialleleR"</span><span class="op">)</span></span>
<span><span class="va">amplicon.bed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"amplicon.bed"</span>, package<span class="op">=</span><span class="st">"epialleleR"</span><span class="op">)</span></span>
<span><span class="va">amplicon.vcf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"amplicon.vcf.gz"</span>, package<span class="op">=</span><span class="st">"epialleleR"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># preloading the data</span></span>
<span><span class="va">bam.data</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/preprocessBam.html">preprocessBam</a></span><span class="op">(</span><span class="va">amplicon.bam</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># methylation patterns, check vignettes or method description for plotting them</span></span>
<span><span class="va">patterns</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/extractPatterns.html">extractPatterns</a></span><span class="op">(</span>bam<span class="op">=</span><span class="va">amplicon.bam</span>, bed<span class="op">=</span><span class="va">amplicon.bed</span>, bed.row<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># CpG VEF report for individual bases</span></span>
<span><span class="va">cg.vef.report</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/generateCytosineReport.html">generateCytosineReport</a></span><span class="op">(</span><span class="va">bam.data</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># BED-guided VEF report for genomic ranges</span></span>
<span><span class="va">bed.report</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/generateBedReport.html">generateBedReport</a></span><span class="op">(</span>bam<span class="op">=</span><span class="va">amplicon.bam</span>, bed<span class="op">=</span><span class="va">amplicon.bed</span>,</span>
<span>                                bed.type<span class="op">=</span><span class="st">"capture"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># VCF report</span></span>
<span><span class="va">vcf.report</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/generateVcfReport.html">generateVcfReport</a></span><span class="op">(</span>bam<span class="op">=</span><span class="va">amplicon.bam</span>, bed<span class="op">=</span><span class="va">amplicon.bed</span>,</span>
<span>                                vcf<span class="op">=</span><span class="va">amplicon.vcf</span>, vcf.style<span class="op">=</span><span class="st">"NCBI"</span><span class="op">)</span></span></code></pre></div>
<hr>
<div class="section level3">
<h3 id="citing-the-epialleler-package">Citing the <em><code>epialleleR</code></em> package<a class="anchor" aria-label="anchor" href="#citing-the-epialleler-package"></a>
</h3>
<p>Oleksii Nikolaienko, Per Eystein Lønning, Stian Knappskog, <em>epialleleR</em>: an R/Bioconductor package for sensitive allele-specific methylation analysis in NGS data. <em>bioRxiv</em> 2022.06.30.498213. <a href="https://www.biorxiv.org/content/10.1101/2022.06.30.498213" class="external-link">https://www.biorxiv.org/content/10.1101/2022.06.30.498213</a></p>
</div>
<div class="section level3">
<h3 id="the-experimental-data-analysed-using-the-package">The experimental data analysed using the package<a class="anchor" aria-label="anchor" href="#the-experimental-data-analysed-using-the-package"></a>
</h3>
<p>Per Eystein Lonning, Oleksii Nikolaienko, Kathy Pan, Allison W. Kurian, Hans Petter Petter Eikesdal, Mary Pettinger, Garnet L Anderson, Ross L Prentice, Rowan T. Chlebowski, and Stian Knappskog. Constitutional <em>BRCA1</em> methylation and risk of incident triple-negative breast cancer and high-grade serous ovarian cancer. <em>JAMA Oncology</em> 2022. <a href="https://doi.org/10.1001/jamaoncol.2022.3846" class="external-link">https://doi.org/10.1001/jamaoncol.2022.3846</a></p>
<p>Oleksii Nikolaienko, Hans P. Eikesdal, Bjørnar Gilje, Steinar Lundgren, Egil S. Blix, Helge Espelid, Jürgen Geisler, Stephanie Geisler, Emiel A.M. Janssen, Synnøve Yndestad, Laura Minsaas, Beryl Leirvaag, Reidun Lillestøl, Stian Knappskog, Per E. Lønning. Prenatal <em>BRCA1</em> epimutations contribute significantly to triple-negative breast cancer development. <em>medRxiv</em> 2023.05.14.23289949. <a href="https://www.medrxiv.org/content/10.1101/2023.05.14.23289949" class="external-link">https://www.medrxiv.org/content/10.1101/2023.05.14.23289949</a></p>
</div>
<div class="section level3">
<h3 id="epialleler-at-bioconductor">
<em><code>epialleleR</code></em> at Bioconductor<a class="anchor" aria-label="anchor" href="#epialleler-at-bioconductor"></a>
</h3>
<p><a href="https://bioconductor.org/packages/release/bioc/html/epialleleR.html" class="external-link">release</a>, <a href="https://bioconductor.org/packages/devel/bioc/html/epialleleR.html" class="external-link">development version</a></p>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="license">License<a class="anchor" aria-label="anchor" href="#license"></a>
</h2>
<p>Artistic License 2.0</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://www.bioconductor.org/packages/epialleleR" class="external-link">View on Bioconductor</a></li>
<li><a href="https://github.com/BBCG/epialleleR/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/BBCG/epialleleR/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/Artistic-2.0" class="external-link">Artistic-2.0</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing epialleleR</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Oleksii Nikolaienko <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-5910-4934" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Oleksii Nikolaienko.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
