<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Fast, Epiallele-Aware Methylation Caller and Reporter • epialleleR</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Fast, Epiallele-Aware Methylation Caller and Reporter">
<meta name="description" content="Epialleles are specific DNA methylation patterns that are mitotically and/or meiotically inherited. This package calls and reports cytosine methylation as well as frequencies of hypermethylated epialleles at the level of genomic regions or individual cytosines in next-generation sequencing data using binary alignment map (BAM) files as an input. Among other things, this package can also extract and visualise methylation patterns and assess allele specificity of methylation.">
<meta property="og:description" content="Epialleles are specific DNA methylation patterns that are mitotically and/or meiotically inherited. This package calls and reports cytosine methylation as well as frequencies of hypermethylated epialleles at the level of genomic regions or individual cytosines in next-generation sequencing data using binary alignment map (BAM) files as an input. Among other things, this package can also extract and visualise methylation patterns and assess allele specificity of methylation.">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">epialleleR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.13.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="articles/epialleleR.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/values.html">The epialleleR output values</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/BBCG/epialleleR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="fast-epiallele-aware-methylation--caller-and-reporter-">Fast, epiallele-aware methylation<br> caller and reporter <a href="https://github.com/BBCG/epialleleR" class="external-link"><img src="articles/epialleleR_logo.svg" alt="logo" align="right" height="50%" class="pull-right"></a>
<a class="anchor" aria-label="anchor" href="#fast-epiallele-aware-methylation--caller-and-reporter-"></a>
</h1></div>
<p><a href="https://github.com/BBCG/epialleleR/actions" class="external-link"><img src="https://github.com/BBCG/epialleleR/workflows/R-CMD-check-bioc/badge.svg"></a> <a href="https://codecov.io/gh/BBCG/epialleleR/tree/devel" class="external-link"><img src="https://codecov.io/gh/BBCG/epialleleR/branch/devel/graph/badge.svg"></a> <a href="https://bioconductor.org/packages/release/bioc/html/epialleleR.html" class="external-link"><img src="https://bioconductor.org/shields/years-in-bioc/epialleleR.svg"></a> <a href="http://bioconda.github.io/recipes/bioconductor-epialleler/README.html" class="external-link"><img src="https://img.shields.io/badge/install%20with-bioconda-brightgreen.svg?style=flat" alt="install with bioconda"></a></p>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p><em><code>epialleleR</code></em> is an R package for calling and reporting cytosine methylation and hypermethylated variant epiallele frequencies (VEF) at the level of genomic regions or individual cytosines in next-generation sequencing data using binary alignment map (BAM) files as an input. See below for additional functionality.</p>
<div class="section level3">
<h3 id="current-features">Current Features<a class="anchor" aria-label="anchor" href="#current-features"></a>
</h3>
<ul>
<li>calling cytosine methylation and saving calls in BAM file (<em><code>callMethylation</code></em>)</li>
<li>creating sample BAM files given mandatory and optional BAM fields (<em><code>simulateBam</code></em>)</li>
<li>conventional reporting of cytosine methylation (<em><code>generateCytosineReport</code></em>)</li>
<li>reporting the hypermethylated variant epiallele frequency (VEF) at the level of genomic regions (<em><code>generate[Bed|Amplicon|Capture]Report</code></em>) or individual cytosines (<em><code>generateCytosineReport</code></em>)</li>
<li>reporting linearised Methylated Haplotype Load (lMHL, <em><code>generateMhlReport</code></em>)</li>
<li>extracting methylation patterns for genomic region of interest (<em><code>extractPatterns</code></em>)</li>
<li>visualising methylation patterns (<em><code>plotPatterns</code></em>)</li>
<li>testing for the association between epiallele methylation status and sequence variations (<em><code>generateVcfReport</code></em>)</li>
<li>assessing the distribution of per-read beta values for genomic regions of interest (<em><code>generateBedEcdf</code></em>)</li>
</ul>
</div>
<div class="section level3">
<h3 id="recent-improvements">Recent improvements<a class="anchor" aria-label="anchor" href="#recent-improvements"></a>
</h3>
<div class="section level5">
<h5 id="v114-bioc-320">v1.14 [BioC 3.20]<a class="anchor" aria-label="anchor" href="#v114-bioc-320"></a>
</h5>
<ul>
<li>creates pretty plots of methylation patterns</li>
</ul>
</div>
<div class="section level5">
<h5 id="v112-bioc-319">v1.12 [BioC 3.19]<a class="anchor" aria-label="anchor" href="#v112-bioc-319"></a>
</h5>
<ul>
<li>inputs long-read sequencing alignments</li>
<li>full support for short-read sequencing alignments by Illumina DRAGEN, Bismark, bwa-meth, BSMAP</li>
<li>RRBS-specific options</li>
<li>lower memory usage</li>
</ul>
</div>
<div class="section level5">
<h5 id="v110-bioc-318">v1.10 [BioC 3.18]<a class="anchor" aria-label="anchor" href="#v110-bioc-318"></a>
</h5>
<ul>
<li>inputs both single-end and paired-end sequencing alignments</li>
<li>makes and stores methylation calls</li>
<li>creates sample BAM files</li>
<li>reports linearised MHL</li>
</ul>
</div>
<div class="section level5">
<h5 id="v14-bioc-315">v1.4 [BioC 3.15]<a class="anchor" aria-label="anchor" href="#v14-bioc-315"></a>
</h5>
<ul>
<li>significant speed-up</li>
<li>method to extract and visualize methylation patterns</li>
</ul>
</div>
<div class="section level5">
<h5 id="v12-bioc-314">v1.2 [BioC 3.14]<a class="anchor" aria-label="anchor" href="#v12-bioc-314"></a>
</h5>
<ul>
<li>even faster and more memory-efficient BAM loading (by means of HTSlib)</li>
<li>min.baseq parameter to reduce the effect of low quality bases on methylation or SNV calling (in v1.0 the output of <em><code>generateVcfReport</code></em> was equivalent to the one of <code>samtools mpileup -Q 0 ...</code>)</li>
</ul>
<p>check out NEWS for more!</p>
<hr>
</div>
</div>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="section level3">
<h3 id="install-via-bioconductor">install via Bioconductor<a class="anchor" aria-label="anchor" href="#install-via-bioconductor"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"epialleleR"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="install-the-latest-version-via-install_github">Install the latest version via install_github<a class="anchor" aria-label="anchor" href="#install-the-latest-version-via-install_github"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/" class="external-link">devtools</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"BBCG/epialleleR"</span>, build_vignettes<span class="op">=</span><span class="cn">FALSE</span>,</span>
<span>  repos<span class="op">=</span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/repositories.html" class="external-link">repositories</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  dependencies<span class="op">=</span><span class="cn">TRUE</span>, type<span class="op">=</span><span class="st">"source"</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="using-the-package">Using the package<a class="anchor" aria-label="anchor" href="#using-the-package"></a>
</h2>
<p>Please read <em><code>epialleleR</code></em> vignette at <a href="https://bbcg.github.io/epialleleR/articles/epialleleR.html" class="external-link">GitHub pages</a> or within the R environment: <code><a href="articles/epialleleR.html">vignette("epialleleR", package="epialleleR")</a></code>, or consult the function’s help pages for the extensive information on usage, parameters and output values.</p>
<p>Comparison of beta, VEF and lMHL values for various use cases is given by the <a href="https://bbcg.github.io/epialleleR/articles/values.html" class="external-link">values</a> vignette (<code><a href="articles/values.html">vignette("values", package="epialleleR")</a></code>)</p>
<p>Very brief synopsis:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/BBCG/epialleleR" class="external-link">epialleleR</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># make methylation calls if necessary</span></span>
<span><span class="fu"><a href="reference/callMethylation.html">callMethylation</a></span><span class="op">(</span></span>
<span>  input.bam.file<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"test"</span>, <span class="st">"dragen-se-unsort-xg.bam"</span>, package<span class="op">=</span><span class="st">"epialleleR"</span><span class="op">)</span>,</span>
<span>  output.bam.file<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span>pattern<span class="op">=</span><span class="st">"output-"</span>, fileext<span class="op">=</span><span class="st">".bam"</span><span class="op">)</span>,</span>
<span>  genome<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"test"</span>, <span class="st">"reference.fasta.gz"</span>, package<span class="op">=</span><span class="st">"epialleleR"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># make a sample BAM file from scratch</span></span>
<span><span class="fu"><a href="reference/simulateBam.html">simulateBam</a></span><span class="op">(</span>output.bam.file<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span>pattern<span class="op">=</span><span class="st">"simulated-"</span>, fileext<span class="op">=</span><span class="st">".bam"</span><span class="op">)</span>,</span>
<span>            pos<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>, XM<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ZZZzzZZZ"</span>, <span class="st">"ZZzzzzZZ"</span><span class="op">)</span>, XG<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CT"</span>, <span class="st">"AG"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># or use external files</span></span>
<span><span class="va">amplicon.bam</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"amplicon010meth.bam"</span>,</span>
<span>                            package<span class="op">=</span><span class="st">"epialleleR"</span><span class="op">)</span></span>
<span><span class="va">amplicon.bed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"amplicon.bed"</span>, package<span class="op">=</span><span class="st">"epialleleR"</span><span class="op">)</span></span>
<span><span class="va">amplicon.vcf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"amplicon.vcf.gz"</span>, package<span class="op">=</span><span class="st">"epialleleR"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># preload the data</span></span>
<span><span class="va">bam.data</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/preprocessBam.html">preprocessBam</a></span><span class="op">(</span><span class="va">amplicon.bam</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># methylation patterns and their plot</span></span>
<span><span class="va">patterns</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/extractPatterns.html">extractPatterns</a></span><span class="op">(</span>bam<span class="op">=</span><span class="va">amplicon.bam</span>, bed<span class="op">=</span><span class="va">amplicon.bed</span>, bed.row<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/plotPatterns.html">plotPatterns</a></span><span class="op">(</span><span class="va">patterns</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># CpG VEF report for individual bases</span></span>
<span><span class="va">cg.vef.report</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/generateCytosineReport.html">generateCytosineReport</a></span><span class="op">(</span><span class="va">bam.data</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># BED-guided VEF report for genomic ranges</span></span>
<span><span class="va">bed.report</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/generateBedReport.html">generateBedReport</a></span><span class="op">(</span>bam<span class="op">=</span><span class="va">amplicon.bam</span>, bed<span class="op">=</span><span class="va">amplicon.bed</span>,</span>
<span>                                bed.type<span class="op">=</span><span class="st">"capture"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># VCF report</span></span>
<span><span class="va">vcf.report</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/generateVcfReport.html">generateVcfReport</a></span><span class="op">(</span>bam<span class="op">=</span><span class="va">amplicon.bam</span>, bed<span class="op">=</span><span class="va">amplicon.bed</span>,</span>
<span>                                vcf<span class="op">=</span><span class="va">amplicon.vcf</span>, vcf.style<span class="op">=</span><span class="st">"NCBI"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># lMHL report</span></span>
<span><span class="va">mhl.report</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/generateMhlReport.html">generateMhlReport</a></span><span class="op">(</span>bam<span class="op">=</span><span class="va">amplicon.bam</span><span class="op">)</span></span></code></pre></div>
<hr>
<div class="section level3">
<h3 id="citing-the-epialleler-package">Citing the <em><code>epialleleR</code></em> package<a class="anchor" aria-label="anchor" href="#citing-the-epialleler-package"></a>
</h3>
<p>Oleksii Nikolaienko, Per Eystein Lønning, Stian Knappskog, <em>epialleleR</em>: an R/Bioconductor package for sensitive allele-specific methylation analysis in NGS data. <em>GigaScience</em>, Volume 12, 2023, giad087, <a href="https://doi.org/10.1093/gigascience/giad087" class="external-link">https://doi.org/10.1093/gigascience/giad087</a>. Data: <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE201690" class="external-link">GSE201690</a></p>
</div>
<div class="section level3">
<h3 id="our-experimental-studies-that-use-the-package">Our experimental studies that use the package<a class="anchor" aria-label="anchor" href="#our-experimental-studies-that-use-the-package"></a>
</h3>
<p>Per Eystein Lonning, Oleksii Nikolaienko, Kathy Pan, Allison W. Kurian, Hans Petter Petter Eikesdal, Mary Pettinger, Garnet L Anderson, Ross L Prentice, Rowan T. Chlebowski, and Stian Knappskog. Constitutional <em>BRCA1</em> methylation and risk of incident triple-negative breast cancer and high-grade serous ovarian cancer. <em>JAMA Oncology</em> 2022. <a href="https://doi.org/10.1001/jamaoncol.2022.3846" class="external-link">https://doi.org/10.1001/jamaoncol.2022.3846</a></p>
<p>Oleksii Nikolaienko, Hans P. Eikesdal, Elisabet Ognedal, Bjørnar Gilje, Steinar Lundgren, Egil S. Blix, Helge Espelid, Jürgen Geisler, Stephanie Geisler, Emiel A.M. Janssen, Synnøve Yndestad, Laura Minsaas, Beryl Leirvaag, Reidun Lillestøl, Stian Knappskog, Per E. Lønning. Prenatal <em>BRCA1</em> epimutations contribute significantly to triple-negative breast cancer development. <em>Genome Medicine</em> 2023. <a href="https://doi.org/10.1186/s13073-023-01262-8" class="external-link">https://doi.org/10.1186/s13073-023-01262-8</a>. Data: <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE243966" class="external-link">GSE243966</a></p>
</div>
<div class="section level3">
<h3 id="epialleler-at-bioconductor">
<em><code>epialleleR</code></em> at Bioconductor<a class="anchor" aria-label="anchor" href="#epialleler-at-bioconductor"></a>
</h3>
<p><a href="https://bioconductor.org/packages/release/bioc/html/epialleleR.html" class="external-link">release</a>, <a href="https://bioconductor.org/packages/devel/bioc/html/epialleleR.html" class="external-link">development version</a></p>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="license">License<a class="anchor" aria-label="anchor" href="#license"></a>
</h2>
<p>Artistic License 2.0</p>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://www.bioconductor.org/packages/epialleleR" class="external-link">View on Bioconductor</a></li>
<li><a href="https://github.com/BBCG/epialleleR/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/BBCG/epialleleR/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/Artistic-2.0" class="external-link">Artistic-2.0</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing epialleleR</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Oleksii Nikolaienko <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-5910-4934" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Oleksii Nikolaienko.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

    </footer>
</div>





  </body>
</html>
