<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Fast, Epiallele-Aware Methylation Reporter • epialleleR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Fast, Epiallele-Aware Methylation Reporter">
<meta property="og:description" content="Epialleles are specific DNA methylation patterns that are
  mitotically and/or meiotically inherited. This package calls hypermethylated
  epiallele frequencies at the level of genomic regions or individual cytosines
  in next-generation sequencing data using binary alignment map (BAM) files as
  an input. Other functionality includes computing the empirical cumulative
  distribution function for per-read beta values, and testing the significance
  of the association between epiallele methylation status and base frequencies
  at particular genomic positions (SNPs).">
<meta property="og:image" content="/logo.svg">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">epialleleR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="articles/epialleleR.html">Get started</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/BBCG/epialleleR/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="fast-epiallele-aware-methylation-reporter-" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#fast-epiallele-aware-methylation-reporter-" class="anchor"></a>Fast, epiallele-aware methylation reporter <img align="right" src="logo.svg">
</h1></div>
<div class="figure">
<img src="https://img.shields.io/github/workflow/status/BBCG/epialleleR/R-CMD-check-bioc" alt=""><p class="caption">R-CMD-check-bioc</p>
</div>
<!-- <br><br><br> -->
<!-- ![epialleleR](logo.svg) &nbsp;
========== -->
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p>DISCLAIMER: This is a work in progress, however the package is already usable, and the obtained experimental results will soon be published. Main methods (<em><code>preprocessBam</code></em>, <em><code>generateCytosineReport</code></em>, <em><code>generateBedReport</code></em>) won’t change. The <em><code>generateVcfReport</code></em> method will at some point be improved to include variable-length sequence variations, while <em><code>generateBedEcdf</code></em> should be considered somewhat experimental and may undergo significant changes or be substituted with some other method in the future.</p>
<p><em><code>epialleleR</code></em> is an R package for calling hypermethylated variant epiallele frequencies (VEF) at the level of genomic regions or individual cytosines in next-generation sequencing data using binary alignment map (BAM) files as an input. Other functionality includes computing the empirical cumulative distribution function for per-read beta values, and testing the significance of the association between epiallele methylation status and base frequencies at particular genomic positions (SNPs).</p>
<div id="current-features" class="section level3">
<h3 class="hasAnchor">
<a href="#current-features" class="anchor"></a>Current Features</h3>
<ul>
<li>conventional reporting of cytosine methylation (<em><code>generateCytosineReport</code></em>)</li>
<li>calling the hypermethylated variant epiallele frequency (VEF) at the level of genomic regions (<em><code>generateBedReport</code></em>) or individual cytosines (<em><code>generateCytosineReport</code></em>)</li>
<li>assessing the distribution of per-read beta values for genomic regions of interest (<em><code>generateBedEcdf</code></em>)</li>
<li>testing for the association between epiallele methylation status and sequence variations (<em><code>generateVcfReport</code></em>)</li>
</ul>
<hr>
</div>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<div id="install-via-bioconductor-not-yet-available" class="section level3">
<h3 class="hasAnchor">
<a href="#install-via-bioconductor-not-yet-available" class="anchor"></a>install via Bioconductor (NOT YET AVAILABLE)</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>

<span class="co"># The following initializes usage of Bioc devel</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span>version<span class="op">=</span><span class="st">'devel'</span><span class="op">)</span>

<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"epialleleR"</span><span class="op">)</span></code></pre></div>
</div>
<div id="install-the-latest-version-via-install_github" class="section level3">
<h3 class="hasAnchor">
<a href="#install-the-latest-version-via-install_github" class="anchor"></a>Install the latest version via install_github</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/" class="external-link">devtools</a></span><span class="op">)</span>
<span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"BBCG/epialleleR"</span>, build_vignettes<span class="op">=</span><span class="cn">FALSE</span>,
  repos<span class="op">=</span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/repositories.html" class="external-link">repositories</a></span><span class="op">(</span><span class="op">)</span>,
  dependencies<span class="op">=</span><span class="cn">TRUE</span>, type<span class="op">=</span><span class="st">"source"</span><span class="op">)</span></code></pre></div>
<hr>
</div>
</div>
<div id="using-the-package" class="section level2">
<h2 class="hasAnchor">
<a href="#using-the-package" class="anchor"></a>Using the package</h2>
<p>Please read <em><code>epialleleR</code></em> vignettes at <a href="https://bbcg.github.io/epialleleR/articles/epialleleR.html" class="external-link">GitHub pages</a> or within the R environment: <code><a href="articles/epialleleR.html">vignette("epialleleR", package="epialleleR")</a></code>, or consult the function’s help pages for the extensive information on usage, parameters and output values.</p>
<p>Very brief synopsis:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/BBCG/epialleleR" class="external-link">epialleleR</a></span><span class="op">)</span>

<span class="co"># external files</span>
<span class="va">amplicon.bam</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"amplicon010meth.bam"</span>, package<span class="op">=</span><span class="st">"epialleleR"</span><span class="op">)</span>
<span class="va">amplicon.bed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"amplicon.bed"</span>, package<span class="op">=</span><span class="st">"epialleleR"</span><span class="op">)</span>
<span class="va">amplicon.vcf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"amplicon.vcf.gz"</span>, package<span class="op">=</span><span class="st">"epialleleR"</span><span class="op">)</span>

<span class="co"># preloading the data</span>
<span class="va">bam.data</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/preprocessBam.html">preprocessBam</a></span><span class="op">(</span><span class="va">amplicon.bam</span><span class="op">)</span>

<span class="co"># CpG VEF report for individual bases</span>
<span class="va">cg.vef.report</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/generateCytosineReport.html">generateCytosineReport</a></span><span class="op">(</span><span class="va">bam.data</span><span class="op">)</span>

<span class="co"># BED-guided VEF report for genomic ranges</span>
<span class="va">bed.report</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/generateBedReport.html">generateBedReport</a></span><span class="op">(</span>bam<span class="op">=</span><span class="va">amplicon.bam</span>, bed<span class="op">=</span><span class="va">amplicon.bed</span>, bed.type<span class="op">=</span><span class="st">"capture"</span><span class="op">)</span>

<span class="co"># VCF report</span>
<span class="va">vcf.report</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/generateVcfReport.html">generateVcfReport</a></span><span class="op">(</span>bam<span class="op">=</span><span class="va">amplicon.bam</span>, bed<span class="op">=</span><span class="va">amplicon.bed</span>,
                                vcf<span class="op">=</span><span class="va">amplicon.vcf</span>, vcf.style<span class="op">=</span><span class="st">"NCBI"</span><span class="op">)</span></code></pre></div>
<hr>
<div id="citing-the-epialleler-package" class="section level3">
<h3 class="hasAnchor">
<a href="#citing-the-epialleler-package" class="anchor"></a>Citing the <em><code>epialleleR</code></em> package</h3>
<p>The experimental data analysed using the package has not been published yet. The citation information will be updated in the nearest future.</p>
</div>
<div id="epialleler-at-bioconductor" class="section level3">
<h3 class="hasAnchor">
<a href="#epialleler-at-bioconductor" class="anchor"></a><em><code>epialleleR</code></em> at Bioconductor</h3>
<p>On its way now.</p>
<hr>
</div>
</div>
<div id="license" class="section level2">
<h2 class="hasAnchor">
<a href="#license" class="anchor"></a>License</h2>
<p>Artistic License/GPL</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/BBCG/epialleleR/">https://​github.com/​BBCG/​epialleleR/​</a>
</li>
<li>Report a bug at <br><a href="https://github.com/BBCG/epialleleR/issues">https://​github.com/​BBCG/​epialleleR/​issues</a>
</li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/Artistic-2.0">Artistic-2.0</a></li>
</ul>
</div>


<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Oleksii Nikolaienko <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-5910-4934" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Oleksii Nikolaienko.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 1.6.1.9001.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
