<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>callMethylation — callMethylation • epialleleR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="callMethylation — callMethylation"><meta name="description" content="This function calls cytosine methylation and stores calls in BAM files."><meta property="og:description" content="This function calls cytosine methylation and stores calls in BAM files."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">epialleleR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.13.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/epialleleR.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/values.html">The epialleleR output values &lt;img style="float: right;" src="epialleleR_logo.svg"&gt; </a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/BBCG/epialleleR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>callMethylation</h1>
      <small class="dont-index">Source: <a href="https://github.com/BBCG/epialleleR/blob/devel/R/callMethylation.R" class="external-link"><code>R/callMethylation.R</code></a></small>
      <div class="d-none name"><code>callMethylation.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function calls cytosine methylation and stores calls in BAM files.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">callMethylation</span><span class="op">(</span></span>
<span>  <span class="va">input.bam.file</span>,</span>
<span>  <span class="va">output.bam.file</span>,</span>
<span>  <span class="va">genome</span>,</span>
<span>  nthreads <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-input-bam-file">input.bam.file<a class="anchor" aria-label="anchor" href="#arg-input-bam-file"></a></dt>
<dd><p>input BAM file location string.</p></dd>


<dt id="arg-output-bam-file">output.bam.file<a class="anchor" aria-label="anchor" href="#arg-output-bam-file"></a></dt>
<dd><p>output BAM file location string.</p></dd>


<dt id="arg-genome">genome<a class="anchor" aria-label="anchor" href="#arg-genome"></a></dt>
<dd><p>reference (genomic) sequences file location string or an
output of <code><a href="preprocessGenome.html">preprocessGenome</a></code>.</p></dd>


<dt id="arg-nthreads">nthreads<a class="anchor" aria-label="anchor" href="#arg-nthreads"></a></dt>
<dd><p>non-negative integer for the number of additional HTSlib
threads to be used during file decompression (default: 1).</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>boolean to report progress and timings (default: TRUE).</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>list object with simple statistics of processed ("nrecs") records
and calls made ("ncalled"). Even though "ncalled" can be less than "nrecs"
(e.g., because not all reads are mapped), all records from the input BAM are
written to the output BAM.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The function makes cytosine methylation calls for short-read methylation
(bisulfite/enzymatic) sequencing alignments from input BAM file and writes
them in the XM tag of the output BAM file. Calls are made on the basis of
reference (e.g., genomic) sequence and observed sequence and cytosine context
of reads. Data reading/processing is done by means of HTSlib,
therefore it is possible to significantly (&gt;5x) speed up the calling
using several (4-8) HTSlib decompression threads.</p>
<p>Methylation calling with this function is only possible for sequencing data
obtained using either bisulfite or other similar sequencing method
(enzymatic methylation sequencing). Cytosine methylation in long-read,
native DNA sequencing alignments should be called using other, appropriate
tools.</p>
<p>It is a requirement that the genomic strand the read was aligned
to is known. This information is typically stored in XG tag of
Bismark/Illumina BAM files, or in YD tag of BWA-meth alignment files, or in
ZS tag of BSMAP alignment files.
`epialleleR` is aware of that and will use the whichever tag is available.</p>
<p>The sequence context of cytosines (h/H for "CHH", x/X for "CHG",
z/Z for "CG") is determined based on the
actual (observed) sequence of the read. E.g., if read "ACGT" was aligned
to the forward strand of reference sequence "ACaaGT" with the CIGAR string
"2M2D2M" (2 bases match,
2 reference bases are deleted, 2 bases match), then methylation call string
will be ".Z.." (in contrast to the reference's one of ".H...."). This
makes cytosine calls nearly identical to ones produced by Bismark Bisulfite
Read Mapper and Methylation Caller or Illumina DRAGEN Bio IT Platform,
however with one important distinction: `epialleleR` reports sequence context
of cytosines followed by unknown bases ("CNN") as "H.." instead of "U.."
(unknown; as for example Illumina DRAGEN Bio IT Platform does). Similarly,
forward strand context of "CNG" is reported as "X..", forward strand context
of "CGN" -&gt; "Z..", reverse strand context of "NNG" -&gt; "..H", reverse strand
context of "CNG" -&gt; "..X", reverse strand context of "NCG" -&gt; "..Z".
Both lowercase and uppercase ACGTN symbols in reference sequence are allowed
and correctly recognised, however all the other symbols (e.g., extended IUPAC
symbols, MRSVWYHKDB) within sequences are converted to N.</p>
<p>As a reference sequence, the function expects either location of
(preferably `bgzip`ped) FASTA file or an object obtained by
<code><a href="preprocessGenome.html">preprocessGenome</a></code>. The latter is recommended if methylation
calling is to be performed on multiple BAM files.</p>
<p>The alignment records of the output BAM file will contain additional XM tag
with the methylation
call string for every mapped read which did not have XM tag available.
Besides that, XG tag with reference sequence strand ("CT" or "GA") is added
to such reads in case it wasn't present.</p>
<p>Please note that for the purpose of methylation calling, the very same
reference genome must be used for both alignment (when BAM is produced) and
calling cytosine methylation by <code>callMethylation</code> method.
Exception is thrown if reference sequence header of BAM file doesn't match
reference sequence data provided (this matching is performed on the basis
of names and lengths of reference sequences).</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="preprocessGenome.html">preprocessGenome</a></code> for preloading reference sequences
and `epialleleR` vignettes for the description of usage and sample data.</p>
<p><a href="https://www.bioinformatics.babraham.ac.uk/projects/bismark/" class="external-link">Bismark</a> Bisulfite Read Mapper and Methylation Caller,
<a href="https://arxiv.org/abs/1401.1129" class="external-link">bwa-meth</a> for fast and accurate alignment of long bisulfite-seq reads,
<a href="https://doi.org/10.1186/1471-2105-10-232" class="external-link">BSMAP</a>: whole genome bisulfite sequence MAPping program,
or info on <a href="https://support-docs.illumina.com/SW/dragen_v42/Content/SW/DRAGEN/MPipelineMeth_fDG.htm" class="external-link">Illumina DRAGEN Bio IT Platform</a>.</p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span>  <span class="fu">callMethylation</span><span class="op">(</span></span></span>
<span class="r-in"><span>    input.bam.file<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"test"</span>, <span class="st">"dragen-se-unsort-xg.bam"</span>, package<span class="op">=</span><span class="st">"epialleleR"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    output.bam.file<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span>pattern<span class="op">=</span><span class="st">"output-"</span>, fileext<span class="op">=</span><span class="st">".bam"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    genome<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"test"</span>, <span class="st">"reference.fasta.gz"</span>, package<span class="op">=</span><span class="st">"epialleleR"</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading reference genome file </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [0.000s]</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Making methylation calls </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [0.022s]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $nrecs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 100</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $ncalled</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 100</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Oleksii Nikolaienko.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

    </footer></div>





  </body></html>

