<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>extractPatterns — extractPatterns • epialleleR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="extractPatterns — extractPatterns"><meta name="description" content="This function extracts methylation patterns (epialleles) for a given genomic
region of interest."><meta property="og:description" content="This function extracts methylation patterns (epialleles) for a given genomic
region of interest."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">epialleleR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.13.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/epialleleR.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/values.html">The epialleleR output values</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/BBCG/epialleleR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>extractPatterns</h1>
      <small class="dont-index">Source: <a href="https://github.com/BBCG/epialleleR/blob/devel/R/extractPatterns.R" class="external-link"><code>R/extractPatterns.R</code></a></small>
      <div class="d-none name"><code>extractPatterns.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function extracts methylation patterns (epialleles) for a given genomic
region of interest.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">extractPatterns</span><span class="op">(</span></span>
<span>  <span class="va">bam</span>,</span>
<span>  <span class="va">bed</span>,</span>
<span>  bed.row <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  zero.based.bed <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  match.min.overlap <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  extract.context <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CG"</span>, <span class="st">"CHG"</span>, <span class="st">"CHH"</span>, <span class="st">"CxG"</span>, <span class="st">"CX"</span><span class="op">)</span>,</span>
<span>  min.context.freq <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  clip.patterns <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  strand.offset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>CG <span class="op">=</span> <span class="fl">1</span>, CHG <span class="op">=</span> <span class="fl">2</span>, CHH <span class="op">=</span> <span class="fl">0</span>, CxG <span class="op">=</span> <span class="fl">0</span>, CX <span class="op">=</span> <span class="fl">0</span><span class="op">)</span><span class="op">[</span><span class="va">extract.context</span><span class="op">]</span>,</span>
<span>  highlight.positions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  <span class="va">...</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-bam">bam<a class="anchor" aria-label="anchor" href="#arg-bam"></a></dt>
<dd><p>BAM file location string OR preprocessed output of
<code><a href="preprocessBam.html">preprocessBam</a></code> function. Read more about BAM file
requirements and BAM preprocessing at <code><a href="preprocessBam.html">preprocessBam</a></code>.</p></dd>


<dt id="arg-bed">bed<a class="anchor" aria-label="anchor" href="#arg-bed"></a></dt>
<dd><p>Browser Extensible Data (BED) file location string OR object of
class <code>GRanges</code> holding genomic coordinates for
regions of interest. It is used to match sequencing reads to the genomic
regions for pattern extraction. The style of seqlevels of BED file/object
must match the style of seqlevels of the BAM file/object used. The
BED/<code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></code> rows are <strong>not</strong> sorted
internally. As of now, the strand information is ignored and patterns
matching both strands are extracted.</p></dd>


<dt id="arg-bed-row">bed.row<a class="anchor" aria-label="anchor" href="#arg-bed-row"></a></dt>
<dd><p>single non-negative integer specifying what `bed` region
should be included in the output (default: 1).</p></dd>


<dt id="arg-zero-based-bed">zero.based.bed<a class="anchor" aria-label="anchor" href="#arg-zero-based-bed"></a></dt>
<dd><p>boolean defining if BED coordinates are zero based
(default: FALSE).</p></dd>


<dt id="arg-match-min-overlap">match.min.overlap<a class="anchor" aria-label="anchor" href="#arg-match-min-overlap"></a></dt>
<dd><p>integer for the smallest overlap between read's and
BED/<code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></code> start or end positions during
matching of capture-based NGS reads (default: 1).</p></dd>


<dt id="arg-extract-context">extract.context<a class="anchor" aria-label="anchor" href="#arg-extract-context"></a></dt>
<dd><p>string defining cytosine methylation context used
to report:</p><ul><li><p>"CG" (the default) – CpG cytosines (called as zZ)</p></li>
<li><p>"CHG" – CHG cytosines (xX)</p></li>
<li><p>"CHH" – CHH cytosines (hH)</p></li>
<li><p>"CxG" – CG and CHG cytosines (zZxX)</p></li>
<li><p>"CX" – all cytosines</p></li>
</ul></dd>


<dt id="arg-min-context-freq">min.context.freq<a class="anchor" aria-label="anchor" href="#arg-min-context-freq"></a></dt>
<dd><p>real number in the range [0;1] (default: 0.01).
Genomic positions that are covered by smaller fraction of patterns (e.g.,
with erroneous context) won't be included in the report.</p></dd>


<dt id="arg-clip-patterns">clip.patterns<a class="anchor" aria-label="anchor" href="#arg-clip-patterns"></a></dt>
<dd><p>boolean if patterns should not extend over the edge of
`bed` region (default: FALSE).</p></dd>


<dt id="arg-strand-offset">strand.offset<a class="anchor" aria-label="anchor" href="#arg-strand-offset"></a></dt>
<dd><p>single non-negative integer specifying the offset of
bases at the reverse (-) strand compared to the forward (+) strand. Allows
to "merge" genomic positions when methylation is symmetric (in CG and CHG
contexts). By default, equals 1 for `extract.context`=="CG", 2 for "CHG", or
0 otherwise.</p></dd>


<dt id="arg-highlight-positions">highlight.positions<a class="anchor" aria-label="anchor" href="#arg-highlight-positions"></a></dt>
<dd><p>integer vector with genomic positions of bases
to include in every overlapping pattern. Allows to visualize the
distribution of single-nucleotide variations (SNVs) among methylation
patterns. `highlight.positions` takes precedence if any of these positions
overlap with within-the-context positions of methylation pattern.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>other parameters to pass to the
<code><a href="preprocessBam.html">preprocessBam</a></code> function.
Options have no effect if preprocessed BAM data was supplied as an input.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>boolean to report progress and timings (default: TRUE).</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p><code><a href="https://rdatatable.gitlab.io/data.table/reference/data.table.html" class="external-link">data.table</a></code> object containing
per-read (pair) base methylation information for the genomic region of
interest. The report columns are:</p><ul><li><p>seqnames – read (pair) reference sequence name</p></li>
<li><p>strand – read (pair) strand</p></li>
<li><p>start – start of the read (pair)</p></li>
<li><p>end – end of the read (pair)</p></li>
<li><p>nbase – number of within-the-context bases for this read (pair)</p></li>
<li><p>beta – beta value of this read (pair), calculated as a ratio of the
  number of methylated within-the-context bases to the total number of
  within-the-context bases</p></li>
<li><p>pattern – hex representation of 64-bit FNV-1a hash calculated for
  all reported base positions and bases in this read (pair). This
  hash value depends only on included genomic positions and their methylation
  call string chars (hHxXzZ) or nucleotides (ACGT, for highlighted bases
  only), thus it is expected to be unique for every
  methylation pattern, although equal for identical methylation patterns
  independently on read (pair) start, end, or strand (when correct
  `strand.offset` is given)</p></li>
<li><p>... – columns for each genomic position that hold corresponding
  methylation call string char, or NA if position is not present in the read
  (pair)</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The function matches reads (for paired-end sequencing alignment files - read
pairs as a single entity) to the genomic
region provided in a BED file/<code>GRanges</code> object, extracts
methylation statuses of bases within those reads, and returns a data frame
which can be used for further analysis and/or plotting of DNA methylation
patterns by <code><a href="plotPatterns.html">plotPatterns</a></code> function.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="plotPatterns.html">plotPatterns</a></code> for pretty plotting of the output,
<code><a href="preprocessBam.html">preprocessBam</a></code> for preloading BAM data,
<code><a href="generateCytosineReport.html">generateCytosineReport</a></code> for methylation statistics at the level
of individual cytosines, <code><a href="generateBedReport.html">generateBedReport</a></code> for genomic
region-based statistics, <code><a href="generateVcfReport.html">generateVcfReport</a></code> for evaluating
epiallele-SNV associations, <code><a href="generateBedEcdf.html">generateBedEcdf</a></code> for analysing the
distribution of per-read beta values, and `epialleleR` vignettes for the
description of usage and sample data.</p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span>  <span class="co"># amplicon data</span></span></span>
<span class="r-in"><span>  <span class="va">amplicon.bam</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"amplicon010meth.bam"</span>,</span></span>
<span class="r-in"><span>                              package<span class="op">=</span><span class="st">"epialleleR"</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="va">amplicon.bed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"amplicon.bed"</span>,</span></span>
<span class="r-in"><span>                              package<span class="op">=</span><span class="st">"epialleleR"</span><span class="op">)</span></span></span>
<span class="r-in"><span>  </span></span>
<span class="r-in"><span>  <span class="co"># extract patterns</span></span></span>
<span class="r-in"><span>  <span class="va">patterns</span> <span class="op">&lt;-</span> <span class="fu">extractPatterns</span><span class="op">(</span>bam<span class="op">=</span><span class="va">amplicon.bam</span>, bed<span class="op">=</span><span class="va">amplicon.bed</span>, bed.row<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading BED file </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [0.033s]</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Checking BAM file: </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> short-read, paired-end, name-sorted alignment detected</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading paired-end BAM file </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [0.009s]</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Extracting methylation patterns </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [0.026s]</span>
<span class="r-in"><span>  </span></span>
<span class="r-in"><span>  <span class="co"># and then plot them</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="plotPatterns.html">plotPatterns</a></span><span class="op">(</span><span class="va">patterns</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 238 patterns supplied</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 45 unique</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 9 most frequent unique patterns were selected for plotting using 10 beta value bins:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [0,0.1) [0.1,0.2) [0.2,0.3) [0.3,0.4) [0.4,0.5) [0.5,0.6) [0.6,0.7) [0.7,0.8) [0.8,0.9) [0.9,1]</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>       2         1         1         0         0         0         0         1         2       2</span>
<span class="r-plt img"><img src="extractPatterns-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span>  </span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Oleksii Nikolaienko.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

    </footer></div>





  </body></html>

