<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>generateBedEcdf — generateBedEcdf • epialleleR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="generateBedEcdf — generateBedEcdf"><meta name="description" content="This function computes empirical cumulative distribution functions (eCDF) for
per-read beta values of the sequencing reads."><meta property="og:description" content="This function computes empirical cumulative distribution functions (eCDF) for
per-read beta values of the sequencing reads."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">epialleleR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.13.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/epialleleR.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/values.html">The epialleleR output values &lt;img style="float: right;" src="epialleleR_logo.svg"&gt; </a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/BBCG/epialleleR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>generateBedEcdf</h1>
      <small class="dont-index">Source: <a href="https://github.com/BBCG/epialleleR/blob/devel/R/generateBedEcdf.R" class="external-link"><code>R/generateBedEcdf.R</code></a></small>
      <div class="d-none name"><code>generateBedEcdf.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function computes empirical cumulative distribution functions (eCDF) for
per-read beta values of the sequencing reads.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">generateBedEcdf</span><span class="op">(</span></span>
<span>  <span class="va">bam</span>,</span>
<span>  <span class="va">bed</span>,</span>
<span>  bed.type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"amplicon"</span>, <span class="st">"capture"</span><span class="op">)</span>,</span>
<span>  bed.rows <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>,</span>
<span>  zero.based.bed <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  match.tolerance <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  match.min.overlap <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  ecdf.context <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CG"</span>, <span class="st">"CHG"</span>, <span class="st">"CHH"</span>, <span class="st">"CxG"</span>, <span class="st">"CX"</span><span class="op">)</span>,</span>
<span>  <span class="va">...</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-bam">bam<a class="anchor" aria-label="anchor" href="#arg-bam"></a></dt>
<dd><p>BAM file location string OR preprocessed output of
<code><a href="preprocessBam.html">preprocessBam</a></code> function. Read more about BAM file
requirements and BAM preprocessing at <code><a href="preprocessBam.html">preprocessBam</a></code>.</p></dd>


<dt id="arg-bed">bed<a class="anchor" aria-label="anchor" href="#arg-bed"></a></dt>
<dd><p>Browser Extensible Data (BED) file location string OR object of
class <code>GRanges</code> holding genomic coordinates for
regions of interest. It is used to match sequencing reads to the genomic
regions prior to eCDF computation. The style of seqlevels of BED file/object
must match the style of seqlevels of the BAM file/object used.</p></dd>


<dt id="arg-bed-type">bed.type<a class="anchor" aria-label="anchor" href="#arg-bed-type"></a></dt>
<dd><p>character string for the type of assay that was used to
produce sequencing reads:</p><ul><li><p>"amplicon" (the default) – used for amplicon-based next-generation
  sequencing when exact coordinates of sequenced fragments are known.
  Matching of reads to genomic ranges are then performed by the read's start
  or end positions, either of which should be no further than
  `match.tolerance` bases away from the start or end position of genomic
  ranges given in BED file/<code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></code> object</p></li>
<li><p>"capture" – used for capture-based next-generation sequencing when
  reads partially overlap with the capture target regions. Read is considered
  to match the genomic range when their overlap is more or equal to
  `match.min.overlap`. If read matches two or more BED genomic regions, only
  the first match is taken (input <code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></code> are
  <strong>not</strong> sorted internally)</p></li>
</ul></dd>


<dt id="arg-bed-rows">bed.rows<a class="anchor" aria-label="anchor" href="#arg-bed-rows"></a></dt>
<dd><p>integer vector specifying what `bed` regions should be
included in the output. If `c(1)` (the default), then function returns eCDFs
for the first region of `bed`, if NULL - eCDF functions for all `bed`
genomic regions as well as for the reads that didn't match any of the regions
(last element of the return value; only if there are such reads).</p></dd>


<dt id="arg-zero-based-bed">zero.based.bed<a class="anchor" aria-label="anchor" href="#arg-zero-based-bed"></a></dt>
<dd><p>boolean defining if BED coordinates are zero based
(default: FALSE).</p></dd>


<dt id="arg-match-tolerance">match.tolerance<a class="anchor" aria-label="anchor" href="#arg-match-tolerance"></a></dt>
<dd><p>integer for the largest difference between read's and
BED <code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></code> start or end positions during
matching of amplicon-based NGS reads (default: 1).</p></dd>


<dt id="arg-match-min-overlap">match.min.overlap<a class="anchor" aria-label="anchor" href="#arg-match-min-overlap"></a></dt>
<dd><p>integer for the smallest overlap between read's and
BED <code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></code> start or end positions during
matching of capture-based NGS reads (default: 1). If read matches two or more
BED genomic regions, only the first match is taken (input
<code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></code> are <strong>not</strong> sorted internally).</p></dd>


<dt id="arg-ecdf-context">ecdf.context<a class="anchor" aria-label="anchor" href="#arg-ecdf-context"></a></dt>
<dd><p>string defining cytosine methylation context used
for computing within-the-context and out-of-context eCDFs:</p><ul><li><p>"CG" (the default) – within-the-context: CpG cytosines (called as
  zZ), out-of-context: all the other cytosines (hHxX)</p></li>
<li><p>"CHG" – within-the-context: CHG cytosines (xX), out-of-context: hHzZ</p></li>
<li><p>"CHH" – within-the-context: CHH cytosines (hH), out-of-context: xXzZ</p></li>
<li><p>"CxG" – within-the-context: CG and CHG cytosines (zZxX),
  out-of-context: CHH cytosines (hH)</p></li>
<li><p>"CX" – all cytosines are considered within-the-context</p></li>
</ul></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>other parameters to pass to the
<code><a href="preprocessBam.html">preprocessBam</a></code> function.
Options have no effect if preprocessed BAM data was supplied as an input.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>boolean to report progress and timings (default: TRUE).</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>list with a number of elements equal to the length of `bed.rows` (if
not NULL), or to the number of genomic regions within `bed` (if
`bed.rows==NULL`) plus one item for all reads not matching `bed` genomic
regions (if any). Every list item is a list on it's own, consisting of two
eCDF functions for within- and out-of-context per-read beta values.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The function matches reads (for paired-end sequencing alignment files - read
pairs as a single entity) to the genomic
regions provided in a BED file/<code>GRanges</code> object, computes
average per-read beta values according to the cytosine context parameter
`ecdf.context`, and returns a list of eCDFs for within- and out-of-context
average per-read beta values, which can be used for plotting.</p>
<p>The resulting eCDFs and their plots can be used to characterise the
methylation pattern of a particular genomic region, e.g. if reads that match
to that region are methylated in an "all-CpGs-or-none" manner or if some
intermediate methylation levels are more frequent.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="preprocessBam.html">preprocessBam</a></code> for preloading BAM data,
<code><a href="generateCytosineReport.html">generateCytosineReport</a></code> for methylation statistics at the level
of individual cytosines, <code><a href="generateBedReport.html">generateBedReport</a></code> for genomic
region-based statistics, <code><a href="generateVcfReport.html">generateVcfReport</a></code> for evaluating
epiallele-SNV associations, <code><a href="extractPatterns.html">extractPatterns</a></code> for exploring
methylation patterns and <code><a href="plotPatterns.html">plotPatterns</a></code> for pretty plotting
of its output, and `epialleleR` vignettes for the description of
usage and sample data.</p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span>  <span class="co"># amplicon data</span></span></span>
<span class="r-in"><span>  <span class="va">amplicon.bam</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"amplicon010meth.bam"</span>,</span></span>
<span class="r-in"><span>                              package<span class="op">=</span><span class="st">"epialleleR"</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="va">amplicon.bed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"amplicon.bed"</span>,</span></span>
<span class="r-in"><span>                              package<span class="op">=</span><span class="st">"epialleleR"</span><span class="op">)</span></span></span>
<span class="r-in"><span>  </span></span>
<span class="r-in"><span>  <span class="co"># let's compute eCDF</span></span></span>
<span class="r-in"><span>  <span class="va">amplicon.ecdfs</span> <span class="op">&lt;-</span> <span class="fu">generateBedEcdf</span><span class="op">(</span>bam<span class="op">=</span><span class="va">amplicon.bam</span>, bed<span class="op">=</span><span class="va">amplicon.bed</span>,</span></span>
<span class="r-in"><span>                                    bed.rows<span class="op">=</span><span class="cn">NULL</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading BED file </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [0.007s]</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Checking BAM file: </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> short-read, paired-end, name-sorted alignment detected</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading paired-end BAM file </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [0.005s]</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Computing ECDFs for within- and out-of-context per-read beta values </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [0.042s]</span>
<span class="r-in"><span>  </span></span>
<span class="r-in"><span>  <span class="co"># there are 5 items in amplicon.ecdfs, let's plot them all</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">amplicon.ecdfs</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span>  </span></span>
<span class="r-in"><span>  <span class="co"># cycle through items</span></span></span>
<span class="r-in"><span>  <span class="kw">for</span> <span class="op">(</span><span class="va">x</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">amplicon.ecdfs</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span>    <span class="co"># four of them have names corresponding to amplicon.bed genomic regions, </span></span></span>
<span class="r-in"><span>    <span class="co"># fifth - NA for all the reads that don't match to any of those regions</span></span></span>
<span class="r-in"><span>    <span class="va">main</span> <span class="op">&lt;-</span> <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">amplicon.ecdfs</span><span class="op">[</span><span class="va">x</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="st">"unmatched"</span></span></span>
<span class="r-in"><span>            <span class="kw">else</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">amplicon.ecdfs</span><span class="op">[</span><span class="va">x</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-in"><span>    </span></span>
<span class="r-in"><span>    <span class="co"># plotting eCDF for within-the-context per-read beta values (in red)</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">amplicon.ecdfs</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">context</span>, col<span class="op">=</span><span class="st">"red"</span>, verticals<span class="op">=</span><span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>         do.points<span class="op">=</span><span class="cn">FALSE</span>, xlim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>, xlab<span class="op">=</span><span class="st">"per-read beta value"</span>,</span></span>
<span class="r-in"><span>         ylab<span class="op">=</span><span class="st">"cumulative density"</span>, main<span class="op">=</span><span class="va">main</span><span class="op">)</span></span></span>
<span class="r-in"><span>    </span></span>
<span class="r-in"><span>    <span class="co"># adding eCDF for out-of-context per-read beta values (in blue)</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">amplicon.ecdfs</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">out.of.context</span>, add<span class="op">=</span><span class="cn">TRUE</span>, col<span class="op">=</span><span class="st">"blue"</span>,</span></span>
<span class="r-in"><span>         verticals<span class="op">=</span><span class="cn">TRUE</span>, do.points<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="op">}</span></span></span>
<span class="r-plt img"><img src="generateBedEcdf-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span>  </span></span>
<span class="r-in"><span>  <span class="co"># recover default plotting parameters</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span>  </span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Oleksii Nikolaienko.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

    </footer></div>





  </body></html>

