<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>generateBedReport — generateBedReport • epialleleR</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="generateBedReport — generateBedReport" />
<meta property="og:description" content="`generateBedReport`, `generateAmpliconReport`, `generateCaptureReport` --
these functions match BAM reads to the set of genomic locations and return
the fraction of reads with an average methylation level passing an arbitrary
threshold." />
<meta property="og:image" content="/logo.svg" />






<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



<!-- dependencies from examples -->


  

  

  </head>

  <body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">epialleleR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../articles/epialleleR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/BBCG/epialleleR/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>generateBedReport</h1>
    <small class="dont-index">Source: <a href='https://github.com/BBCG/epialleleR/blob/master/R/generateBedReport.R'><code>R/generateBedReport.R</code></a></small>
    <div class="hidden name"><code>generateBedReport.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>`generateBedReport`, `generateAmpliconReport`, `generateCaptureReport` --
these functions match BAM reads to the set of genomic locations and return
the fraction of reads with an average methylation level passing an arbitrary
threshold.</p>
    </div>

    <pre class="usage"><span class='fu'>generateAmpliconReport</span><span class='op'>(</span>
  <span class='va'>bam</span>,
  <span class='va'>bed</span>,
  report.file <span class='op'>=</span> <span class='cn'>NULL</span>,
  zero.based.bed <span class='op'>=</span> <span class='cn'>FALSE</span>,
  match.tolerance <span class='op'>=</span> <span class='fl'>1</span>,
  threshold.reads <span class='op'>=</span> <span class='cn'>TRUE</span>,
  threshold.context <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"CG"</span>, <span class='st'>"CHG"</span>, <span class='st'>"CHH"</span>, <span class='st'>"CxG"</span>, <span class='st'>"CX"</span><span class='op'>)</span>,
  min.context.sites <span class='op'>=</span> <span class='fl'>2</span>,
  min.context.beta <span class='op'>=</span> <span class='fl'>0.5</span>,
  max.outofcontext.beta <span class='op'>=</span> <span class='fl'>0.1</span>,
  min.mapq <span class='op'>=</span> <span class='fl'>0</span>,
  skip.duplicates <span class='op'>=</span> <span class='cn'>FALSE</span>,
  gzip <span class='op'>=</span> <span class='cn'>FALSE</span>,
  verbose <span class='op'>=</span> <span class='cn'>TRUE</span>
<span class='op'>)</span>

<span class='fu'>generateCaptureReport</span><span class='op'>(</span>
  <span class='va'>bam</span>,
  <span class='va'>bed</span>,
  report.file <span class='op'>=</span> <span class='cn'>NULL</span>,
  zero.based.bed <span class='op'>=</span> <span class='cn'>FALSE</span>,
  match.min.overlap <span class='op'>=</span> <span class='fl'>1</span>,
  threshold.reads <span class='op'>=</span> <span class='cn'>TRUE</span>,
  threshold.context <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"CG"</span>, <span class='st'>"CHG"</span>, <span class='st'>"CHH"</span>, <span class='st'>"CxG"</span>, <span class='st'>"CX"</span><span class='op'>)</span>,
  min.context.sites <span class='op'>=</span> <span class='fl'>2</span>,
  min.context.beta <span class='op'>=</span> <span class='fl'>0.5</span>,
  max.outofcontext.beta <span class='op'>=</span> <span class='fl'>0.1</span>,
  min.mapq <span class='op'>=</span> <span class='fl'>0</span>,
  skip.duplicates <span class='op'>=</span> <span class='cn'>FALSE</span>,
  gzip <span class='op'>=</span> <span class='cn'>FALSE</span>,
  verbose <span class='op'>=</span> <span class='cn'>TRUE</span>
<span class='op'>)</span>

<span class='fu'>generateBedReport</span><span class='op'>(</span>
  <span class='va'>bam</span>,
  <span class='va'>bed</span>,
  report.file <span class='op'>=</span> <span class='cn'>NULL</span>,
  zero.based.bed <span class='op'>=</span> <span class='cn'>FALSE</span>,
  bed.type <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"amplicon"</span>, <span class='st'>"capture"</span><span class='op'>)</span>,
  match.tolerance <span class='op'>=</span> <span class='fl'>1</span>,
  match.min.overlap <span class='op'>=</span> <span class='fl'>1</span>,
  threshold.reads <span class='op'>=</span> <span class='cn'>TRUE</span>,
  threshold.context <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"CG"</span>, <span class='st'>"CHG"</span>, <span class='st'>"CHH"</span>, <span class='st'>"CxG"</span>, <span class='st'>"CX"</span><span class='op'>)</span>,
  min.context.sites <span class='op'>=</span> <span class='fl'>2</span>,
  min.context.beta <span class='op'>=</span> <span class='fl'>0.5</span>,
  max.outofcontext.beta <span class='op'>=</span> <span class='fl'>0.1</span>,
  min.mapq <span class='op'>=</span> <span class='fl'>0</span>,
  skip.duplicates <span class='op'>=</span> <span class='cn'>FALSE</span>,
  gzip <span class='op'>=</span> <span class='cn'>FALSE</span>,
  verbose <span class='op'>=</span> <span class='cn'>TRUE</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>bam</th>
      <td><p>BAM file location string OR preprocessed output of
<code><a href='preprocessBam.html'>preprocessBam</a></code> function.</p></td>
    </tr>
    <tr>
      <th>bed</th>
      <td><p>Browser Extensible Data (BED) file location string OR object of
class <code><a href='https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html'>GRanges</a></code> holding genomic coordinates for
regions of interest. The seqlevels of BED file/object must be the same as
seqlevels of BAM file/object used.</p></td>
    </tr>
    <tr>
      <th>report.file</th>
      <td><p>file location string to write the BED report. If NULL
(the default) then report is returned as a
<code><a href='https://Rdatatable.gitlab.io/data.table/reference/data.table.html'>data.table</a></code> object.</p></td>
    </tr>
    <tr>
      <th>zero.based.bed</th>
      <td><p>boolean defining if BED coordinates are zero based
(default: FALSE).</p></td>
    </tr>
    <tr>
      <th>match.tolerance</th>
      <td><p>integer for the largest difference between read's and
BED <code><a href='https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html'>GRanges</a></code> start or end positions during
matching of amplicon-based NGS reads (default: 1).</p></td>
    </tr>
    <tr>
      <th>threshold.reads</th>
      <td><p>boolean defining if sequence reads should be
thresholded before counting reads belonging to variant epialleles (default:
TRUE). Disabling thresholding is possible but makes no sense in this context
as all the reads will be assigned to the variant epiallele, which will result
in VEF==1 (in such case `NA` VEF values are returned in order to avoid
confusion).</p></td>
    </tr>
    <tr>
      <th>threshold.context</th>
      <td><p>string defining cytosine methylation context used
for thresholding the reads:</p><ul>
<li><p>"CG" (the default) -- within-the-context: CpG cytosines (called as
  zZ), out-of-context: all the other cytosines (hHxX)</p></li>
<li><p>"CHG" -- within-the-context: CHG cytosines (xX), out-of-context: hHzZ</p></li>
<li><p>"CHH" -- within-the-context: CHH cytosines (hH), out-of-context: xXzZ</p></li>
<li><p>"CxG" -- within-the-context: CG and CHG cytosines (zZxX),
  out-of-context: CHH cytosines (hH)</p></li>
<li><p>"CX" -- all cytosines are considered within-the-context, this
  effectively results in no thresholding</p></li>
</ul><p>This option has no effect when read thresholding is disabled.</p></td>
    </tr>
    <tr>
      <th>min.context.sites</th>
      <td><p>non-negative integer for minimum number of cytosines
within the `threshold.context` (default: 2). Reads containing <strong>fewer</strong>
within-the-context cytosines are considered completely unmethylated (thus
belonging to the reference epiallele). This option has no effect when read
thresholding is disabled.</p></td>
    </tr>
    <tr>
      <th>min.context.beta</th>
      <td><p>real number in the range [0;1] (default: 0.5). Reads
with average beta value for within-the-context cytosines <strong>below</strong> this
threshold are considered completely unmethylated (thus belonging to the
reference epiallele). This option has no effect when read thresholding is
disabled.</p></td>
    </tr>
    <tr>
      <th>max.outofcontext.beta</th>
      <td><p>real number in the range [0;1] (default: 0.1).
Reads with average beta value for out-of-context cytosines <strong>above</strong>
this threshold are considered completely unmethylated (thus belonging to the
reference epiallele). This option has no effect when read thresholding is
disabled.</p></td>
    </tr>
    <tr>
      <th>min.mapq</th>
      <td><p>non-negative integer threshold for minimum read mapping
quality (default: 0). Option has no effect if preprocessed BAM data was
supplied as an input.</p></td>
    </tr>
    <tr>
      <th>skip.duplicates</th>
      <td><p>boolean defining if duplicate aligned reads should be
skipped (default: FALSE). Option has no effect if preprocessed BAM data was
supplied as an input OR duplicate reads were not marked by alignment
software.</p></td>
    </tr>
    <tr>
      <th>gzip</th>
      <td><p>boolean to compress the report (default: FALSE).</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>boolean to report progress and timings (default: TRUE).</p></td>
    </tr>
    <tr>
      <th>match.min.overlap</th>
      <td><p>integer for the smallest overlap between read's and
BED <code><a href='https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html'>GRanges</a></code> start or end positions during
matching of capture-based NGS reads (default: 1). If read matches two or more
BED genomic regions, only the first match is taken (input
<code><a href='https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html'>GRanges</a></code> are <strong>not</strong> sorted internally).</p></td>
    </tr>
    <tr>
      <th>bed.type</th>
      <td><p>character string for the type of assay that was used to
produce sequencing reads:</p><ul>
<li><p>"amplicon" (the default) -- used for amplicon-based next-generation
  sequencing when exact coordinates of sequenced fragments are known.
  Matching of reads to genomic ranges are then performed by the read's start
  or end positions, either of which should be no further than
  `match.tolerance` bases away from the start or end position of genomic
  ranges given in BED file/<code><a href='https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html'>GRanges</a></code> object</p></li>
<li><p>"capture" -- used for capture-based next-generation sequencing when
  reads partially overlap with the capture target regions. Read is considered
  to match the genomic range when their overlap is more or equal to
  `match.min.overlap`. If read matches two or more BED genomic regions, only
  the first match is taken (input <code><a href='https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html'>GRanges</a></code> are
  <strong>not</strong> sorted internally)</p></li>
</ul></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code><a href='https://Rdatatable.gitlab.io/data.table/reference/data.table.html'>data.table</a></code> object containing VEF report for
BED <code><a href='https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html'>GRanges</a></code> or NULL if report.file was
specified. If BAM file contains reads that would not match to any of BED
<code><a href='https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html'>GRanges</a></code>, the last row in the report will
contain information on such reads (with seqnames, start and end equal to NA).
The report columns are:</p><ul>
<li><p>seqnames -- reference sequence name</p></li>
<li><p>start -- start of genomic region</p></li>
<li><p>end -- end of genomic region</p></li>
<li><p>width -- width of genomic region</p></li>
<li><p>strand -- strand</p></li>
<li><p>... -- other columns that were present in BED or metadata columns of
  <code><a href='https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html'>GRanges</a></code> object</p></li>
<li><p>nreads+ -- number of reads mapped to the forward ("+") strand</p></li>
<li><p>nreads- -- number of reads mapped to the reverse ("-") strand</p></li>
<li><p>VEF -- frequency of reads passing the threshold</p></li>
</ul>

    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Functions report hypermethylated variant epiallele frequencies (VEF) per
genomic region of interest using BAM and BED files as input. Reads (or read
pairs as a single entity) are matched to genomic locations by exact
coordinates (`generateAmpliconReport` or `generateBedReport` with an option
bed.type="amplicon") or minimum overlap (`generateCaptureReport` or
`generateBedReport` with an option bed.type="capture") -- the former to be
used for amplicon-based NGS data, while the latter -- for the capture-based
NGS data. The function's logic is explained below.</p>
<p>Let's suppose we have a BAM file with four reads, all mapped to the "+"
strand of chromosome 1, positions 1-16. The genomic range is supplied as a
parameter `bed = as("chr1:1-100", "GRanges")`. Assuming the default values
for the thresholding parameters (threshold.reads = TRUE,
threshold.context = "CG", min.context.sites = 2, min.context.beta = 0.5,
max.outofcontext.beta = 0.1), the input and results will look as following:</p>
<table class='table'>
<tr><td>methylation string</td><td>threshold</td><td>explained</td></tr>
<tr><td>...Z..x+.h..x..h.</td><td>below</td><td>min.context.sites &lt; 2 (only one zZ base)</td></tr>
<tr><td>...Z..z.h..x..h.</td><td>above</td><td>pass all criteria</td></tr>
<tr><td>...Z..z.h..X..h.</td><td>below</td><td>max.outofcontext.beta &gt; 0.1 (1XH / 3xXhH = 0.33)</td></tr>
<tr><td>...Z..z.h..z-.h.</td><td>below</td><td>min.context.beta &lt; 0.5 (1Z / 3zZ = 0.33)</td></tr>
</table>


<p>Only the second read will satisfy all of the thresholding criteria, leading
to the following BED report (given that all reads map to chr1:+:1-16):</p>
<table class='table'>
<tr><td>seqnames</td><td>start</td><td>end</td><td>width</td><td>strand</td><td>nreads+</td><td>nreads-</td><td>VEF</td></tr>
<tr><td>chr1</td><td>1</td><td>100</td><td>100</td><td>*</td><td>4</td><td>0</td><td>0.25</td></tr>
</table>


    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='preprocessBam.html'>preprocessBam</a></code> for preloading BAM data,
<code><a href='generateCytosineReport.html'>generateCytosineReport</a></code> for methylation statistics at the level
of individual cytosines, <code><a href='generateVcfReport.html'>generateVcfReport</a></code> for evaluating
epiallele-SNV associations, <code><a href='generateBedEcdf.html'>generateBedEcdf</a></code> for analysing the
distribution of per-read beta values, and `epialleleR` vignettes for the
description of usage and sample data.</p>
<p><code><a href='https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html'>GRanges</a></code> class for working with genomic ranges,
<code><a href='https://rdrr.io/pkg/GenomeInfoDb/man/seqlevelsStyle.html'>seqlevelsStyle</a></code> function for getting or setting
the seqlevels style.</p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>  <span class='co'># amplicon data</span>
  <span class='va'>amplicon.bam</span>    <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata"</span>, <span class='st'>"amplicon010meth.bam"</span>, package<span class='op'>=</span><span class='st'>"epialleleR"</span><span class='op'>)</span>
  <span class='va'>amplicon.bed</span>    <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata"</span>, <span class='st'>"amplicon.bed"</span>, package<span class='op'>=</span><span class='st'>"epialleleR"</span><span class='op'>)</span>
  <span class='va'>amplicon.report</span> <span class='op'>&lt;-</span> <span class='fu'>generateAmpliconReport</span><span class='op'>(</span>bam<span class='op'>=</span><span class='va'>amplicon.bam</span>, bed<span class='op'>=</span><span class='va'>amplicon.bed</span><span class='op'>)</span>
</div><div class='co'>#&gt; Reading BED file</div><div class='co'>#&gt;  [0.017s]</div><div class='co'>#&gt; Reading BAM file</div><div class='co'>#&gt;  [0.162s]</div><div class='co'>#&gt; Preprocessing BAM data:</div><div class='co'>#&gt;   Transforming sequences</div><div class='co'>#&gt;  [0.006s]</div><div class='co'>#&gt;   Merging pairs</div><div class='co'>#&gt;  [0.007s]</div><div class='co'>#&gt; Thresholding reads</div><div class='co'>#&gt;  [0.001s]</div><div class='co'>#&gt; Preparing amplicon report</div><div class='co'>#&gt;  [0.027s]</div><div class='input'>  
  <span class='co'># capture NGS</span>
  <span class='va'>capture.bam</span>    <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata"</span>, <span class='st'>"capture.bam"</span>, package<span class='op'>=</span><span class='st'>"epialleleR"</span><span class='op'>)</span>
  <span class='va'>capture.bed</span>    <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata"</span>, <span class='st'>"capture.bed"</span>, package<span class='op'>=</span><span class='st'>"epialleleR"</span><span class='op'>)</span>
  <span class='va'>capture.report</span> <span class='op'>&lt;-</span> <span class='fu'>generateCaptureReport</span><span class='op'>(</span>bam<span class='op'>=</span><span class='va'>capture.bam</span>, bed<span class='op'>=</span><span class='va'>capture.bed</span><span class='op'>)</span>
</div><div class='co'>#&gt; Reading BED file</div><div class='co'>#&gt;  [0.019s]</div><div class='co'>#&gt; Reading BAM file</div><div class='co'>#&gt;  [1.071s]</div><div class='co'>#&gt; Preprocessing BAM data:</div><div class='co'>#&gt;   Transforming sequences</div><div class='co'>#&gt;  [0.020s]</div><div class='co'>#&gt;   Merging pairs</div><div class='co'>#&gt;  [0.020s]</div><div class='co'>#&gt; Thresholding reads</div><div class='co'>#&gt;  [0.002s]</div><div class='co'>#&gt; Preparing capture report</div><div class='co'>#&gt;  [0.046s]</div><div class='input'>  
  <span class='co'># generateAmpliconReport and generateCaptureReport are just aliases</span>
  <span class='co'># of the generateBedReport</span>
  <span class='va'>bed.report</span> <span class='op'>&lt;-</span> <span class='fu'>generateBedReport</span><span class='op'>(</span>bam<span class='op'>=</span><span class='va'>capture.bam</span>, bed<span class='op'>=</span><span class='va'>capture.bed</span>, bed.type<span class='op'>=</span><span class='st'>"capture"</span><span class='op'>)</span>
</div><div class='co'>#&gt; Reading BED file</div><div class='co'>#&gt;  [0.017s]</div><div class='co'>#&gt; Reading BAM file</div><div class='co'>#&gt;  [0.186s]</div><div class='co'>#&gt; Preprocessing BAM data:</div><div class='co'>#&gt;   Transforming sequences</div><div class='co'>#&gt;  [0.017s]</div><div class='co'>#&gt;   Merging pairs</div><div class='co'>#&gt;  [0.015s]</div><div class='co'>#&gt; Thresholding reads</div><div class='co'>#&gt;  [0.002s]</div><div class='co'>#&gt; Preparing capture report</div><div class='co'>#&gt;  [0.019s]</div><div class='input'>  <span class='fu'><a href='https://rdrr.io/r/base/identical.html'>identical</a></span><span class='op'>(</span><span class='va'>capture.report</span>, <span class='va'>bed.report</span><span class='op'>)</span>
</div><div class='co'>#&gt; [1] TRUE</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p><p>Developed by Oleksii Nikolaienko.</p></p>
</div>

<div class="pkgdown">
  <p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 1.6.1.9001.</p></p>
</div>

      </footer>
   </div>

  


  

  </body>
</html>


