<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>generateVcfReport — generateVcfReport • epialleleR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="generateVcfReport — generateVcfReport"><meta name="description" content="This function reports base frequencies at particular genomic positions and
tests their association with the methylation status of the sequencing reads."><meta property="og:description" content="This function reports base frequencies at particular genomic positions and
tests their association with the methylation status of the sequencing reads."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">epialleleR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.13.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/epialleleR.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/values.html">The epialleleR output values</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/BBCG/epialleleR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>generateVcfReport</h1>
      <small class="dont-index">Source: <a href="https://github.com/BBCG/epialleleR/blob/devel/R/generateVcfReport.R" class="external-link"><code>R/generateVcfReport.R</code></a></small>
      <div class="d-none name"><code>generateVcfReport.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function reports base frequencies at particular genomic positions and
tests their association with the methylation status of the sequencing reads.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">generateVcfReport</span><span class="op">(</span></span>
<span>  <span class="va">bam</span>,</span>
<span>  <span class="va">vcf</span>,</span>
<span>  vcf.style <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  bed <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  report.file <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  zero.based.bed <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  threshold.reads <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  threshold.context <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CG"</span>, <span class="st">"CHG"</span>, <span class="st">"CHH"</span>, <span class="st">"CxG"</span>, <span class="st">"CX"</span><span class="op">)</span>,</span>
<span>  min.context.sites <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  min.context.beta <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  max.outofcontext.beta <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  <span class="va">...</span>,</span>
<span>  gzip <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-bam">bam<a class="anchor" aria-label="anchor" href="#arg-bam"></a></dt>
<dd><p>BAM file location string OR preprocessed output of
<code><a href="preprocessBam.html">preprocessBam</a></code> function. Read more about BAM file
requirements and BAM preprocessing at <code><a href="preprocessBam.html">preprocessBam</a></code>.</p></dd>


<dt id="arg-vcf">vcf<a class="anchor" aria-label="anchor" href="#arg-vcf"></a></dt>
<dd><p>Variant Call Format (VCF) file location string OR a VCF object
returned by <code><a href="https://rdrr.io/pkg/VariantAnnotation/man/readVcf-methods.html" class="external-link">readVcf</a></code> function. If VCF object
is supplied, the style of its seqlevels must match the style of seqlevels of
the BAM file/object used.</p></dd>


<dt id="arg-vcf-style">vcf.style<a class="anchor" aria-label="anchor" href="#arg-vcf-style"></a></dt>
<dd><p>string for the seqlevels style of the VCF file, if
different from BED file/object. Only has effect when `vcf` parameter points
to the VCF file location and `bed` is not NULL. Possible values:</p><ul><li><p>NULL (the default) – seqlevels in BED file/object and VCF file are
  the same</p></li>
<li><p>"NCBI", "UCSC", ... – valid parameters of
  <code><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqlevelsStyle.html" class="external-link">seqlevelsStyle</a></code> function</p></li>
</ul></dd>


<dt id="arg-bed">bed<a class="anchor" aria-label="anchor" href="#arg-bed"></a></dt>
<dd><p>Browser Extensible Data (BED) file location string OR object of
class <code>GRanges</code> holding genomic coordinates for
regions of interest. It is used to include only the specific genomic ranges
when the VCF file is loaded. This option has no effect when VCF object is
supplied as a `vcf` parameter. The style of seqlevels of BED file/object
must match the style of seqlevels of the BAM file/object used.</p></dd>


<dt id="arg-report-file">report.file<a class="anchor" aria-label="anchor" href="#arg-report-file"></a></dt>
<dd><p>file location string to write the VCF report. If NULL
(the default) then report is returned as a
<code><a href="https://rdatatable.gitlab.io/data.table/reference/data.table.html" class="external-link">data.table</a></code> object.</p></dd>


<dt id="arg-zero-based-bed">zero.based.bed<a class="anchor" aria-label="anchor" href="#arg-zero-based-bed"></a></dt>
<dd><p>boolean defining if BED coordinates are zero based
(default: FALSE).</p></dd>


<dt id="arg-threshold-reads">threshold.reads<a class="anchor" aria-label="anchor" href="#arg-threshold-reads"></a></dt>
<dd><p>boolean defining if sequence reads should be
thresholded before counting bases in reference and variant epialleles
(default: TRUE). Disabling thresholding is possible but makes no sense in
the context of this function, because
all the reads will be assigned to the variant epiallele,
which will result in Fisher's Exact test p-value of 1 (in columns `FEp+` and
`FEP-`). As thresholding is <strong>not</strong> recommended for long-read
sequencing data, this function is <strong>not</strong> recommended for such data
either.</p></dd>


<dt id="arg-threshold-context">threshold.context<a class="anchor" aria-label="anchor" href="#arg-threshold-context"></a></dt>
<dd><p>string defining cytosine methylation context used
for thresholding the reads:</p><ul><li><p>"CG" (the default) – within-the-context: CpG cytosines (called as
  zZ), out-of-context: all the other cytosines (hHxX)</p></li>
<li><p>"CHG" – within-the-context: CHG cytosines (xX), out-of-context: hHzZ</p></li>
<li><p>"CHH" – within-the-context: CHH cytosines (hH), out-of-context: xXzZ</p></li>
<li><p>"CxG" – within-the-context: CG and CHG cytosines (zZxX),
  out-of-context: CHH cytosines (hH)</p></li>
<li><p>"CX" – all cytosines are considered within-the-context, this
  effectively results in no thresholding</p></li>
</ul><p>This option has no effect when read thresholding is disabled.</p></dd>


<dt id="arg-min-context-sites">min.context.sites<a class="anchor" aria-label="anchor" href="#arg-min-context-sites"></a></dt>
<dd><p>non-negative integer for minimum number of cytosines
within the `threshold.context` (default: 2). Reads containing <strong>fewer</strong>
within-the-context cytosines are considered completely unmethylated (thus
belonging to the reference epiallele). This option has no effect when read
thresholding is disabled.</p></dd>


<dt id="arg-min-context-beta">min.context.beta<a class="anchor" aria-label="anchor" href="#arg-min-context-beta"></a></dt>
<dd><p>real number in the range [0;1] (default: 0.5). Reads
with average beta value for within-the-context cytosines <strong>below</strong> this
threshold are considered completely unmethylated (thus belonging to the
reference epiallele). This option has no effect when read thresholding is
disabled.</p></dd>


<dt id="arg-max-outofcontext-beta">max.outofcontext.beta<a class="anchor" aria-label="anchor" href="#arg-max-outofcontext-beta"></a></dt>
<dd><p>real number in the range [0;1] (default: 0.1).
Reads with average beta value for out-of-context cytosines <strong>above</strong>
this threshold are considered completely unmethylated (thus belonging to the
reference epiallele). This option has no effect when read thresholding is
disabled.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>other parameters to pass to the
<code><a href="preprocessBam.html">preprocessBam</a></code> function.
Options have no effect if preprocessed BAM data was supplied as an input.</p></dd>


<dt id="arg-gzip">gzip<a class="anchor" aria-label="anchor" href="#arg-gzip"></a></dt>
<dd><p>boolean to compress the report (default: FALSE).</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>boolean to report progress and timings (default: TRUE).</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p><code><a href="https://rdatatable.gitlab.io/data.table/reference/data.table.html" class="external-link">data.table</a></code> object containing VCF report or
NULL if report.file was specified. The report columns are:</p><ul><li><p>name – variation identifier (e.g. "rs123456789")</p></li>
<li><p>seqnames – reference sequence name</p></li>
<li><p>range – genomic coordinates of the variation</p></li>
<li><p>REF – base at the reference allele</p></li>
<li><p>ALT – base at the alternative allele</p></li>
<li><p>[M|U][+|-][Ref|Alt] – number of <strong>Ref</strong>erence or
  <strong>Alt</strong>ernative bases that were found at this particular position
  within <strong>M</strong>ethylated (above threshold) or <strong>U</strong>nmethylated
  (below threshold) reads that were mapped to <strong>"+"</strong> (forward)
  or <strong>"-"</strong> (reverse) DNA strand. NA values mean that it is not
  possible to determine the number of bases due to the bisulfite
  conversion-related limitations (C-&gt;T variants on "+" and G-&gt;A on "-"
  strands)</p></li>
<li><p>SumRef – sum of all <strong>Ref</strong>erence base counts</p></li>
<li><p>SumAlt – sum of all <strong>Alt</strong>ernative base counts</p></li>
<li><p>FEp+ – Fisher Exact test p-value for association of a variation
  with methylation status of the reads that map to the <strong>"+"</strong>
  (forward) DNA strand. Calculated using following contingency table:</p><table class="table table"><tr><td>M+Ref</td><td>M+Alt</td></tr><tr><td>U+Ref</td><td>U+Alt</td></tr></table></li>
<li><p>FEp- – Fisher Exact test p-value for association of a variation
  with methylation status of the reads that map to the <strong>"-"</strong>
  (reverse) DNA strand. Calculated using following contingency table:</p><table class="table table"><tr><td>M-Ref</td><td>M-Alt</td></tr><tr><td>U-Ref</td><td>U-Alt</td></tr></table></li>
</ul></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Using BAM reads and sequence variation information as an input,
`generateVcfReport` function thresholds the reads (for paired-end sequencing
alignment files - read pairs as a single
entity) according to supplied parameters and calculates the occurrence of
<strong>Ref</strong>erence and <strong>Alt</strong>ernative bases within reads, taking into
the account DNA strand the read mapped to and average methylation level
(epiallele status) of the read.</p>
<p>The information on sequence variation can be supplied as a Variant Call
Format (VCF) file location or an object of class VCF, returned by the
<code><a href="https://rdrr.io/pkg/VariantAnnotation/man/readVcf-methods.html" class="external-link">readVcf</a></code> function call. As whole-genome VCF
files can be extremely large, it is strongly advised to use only relevant
subset of their data, prefiltering the VCF object manually before calling
`generateVcfReport` or specifying `bed` parameter when `vcf` points to the
location of such large VCF file. Please note that all the BAM, BED and VCF
files must use the same style for seqlevels (i.e. chromosome names).</p>
<p>After counting, function checks if certain bases occur more often within
reads belonging to certain epialleles using Fisher Exact test
(HTSlib's own implementation) and reports separate p-values for reads
mapped to <strong>"+"</strong> (forward) and <strong>"-"</strong> (reverse) DNA strands.</p>
<p>Please note that the final report currently includes only the VCF entries
with single-base REF and ALT alleles. Also, the default (`min.baseq=0`)
output of `generateVcfReport` is equivalent to the one of
`samtools mplieup -Q 0 ...`, and therefore may result in false SNVs caused
by misalignments. Remember to increase `min.baseq` (`samtools mplieup -Q`
default value is 13) to obtain higher-quality results.</p>
<p>Read thresholding by an average methylation level used in this function
makes little sense for long-read sequencing alignments,
as such reads can cover multiple regions with very different DNA methylation
properties. Instead, please use <code><a href="extractPatterns.html">extractPatterns</a></code>,
limiting pattern output to the region of interest only.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="preprocessBam.html">preprocessBam</a></code> for preloading BAM data,
<code><a href="generateCytosineReport.html">generateCytosineReport</a></code> for methylation statistics at the level
of individual cytosines, <code><a href="generateBedReport.html">generateBedReport</a></code> for genomic
region-based statistics, <code><a href="extractPatterns.html">extractPatterns</a></code> for exploring
methylation patterns and <code><a href="plotPatterns.html">plotPatterns</a></code> for pretty plotting
of its output, <code><a href="generateBedEcdf.html">generateBedEcdf</a></code> for analysing the
distribution of per-read beta values, and `epialleleR` vignettes for the
description of usage and sample data.</p>
<p><code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></code> class for working with genomic ranges,
<code><a href="https://rdrr.io/pkg/VariantAnnotation/man/readVcf-methods.html" class="external-link">readVcf</a></code> function for loading VCF data,
<code><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqlevelsStyle.html" class="external-link">seqlevelsStyle</a></code> function for getting or setting
the seqlevels style.</p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span>  <span class="va">capture.bam</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"capture.bam"</span>, package<span class="op">=</span><span class="st">"epialleleR"</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="va">capture.bed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"capture.bed"</span>, package<span class="op">=</span><span class="st">"epialleleR"</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="va">capture.vcf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"capture.vcf.gz"</span>,</span></span>
<span class="r-in"><span>                             package<span class="op">=</span><span class="st">"epialleleR"</span><span class="op">)</span></span></span>
<span class="r-in"><span>  </span></span>
<span class="r-in"><span>  <span class="co"># VCF report</span></span></span>
<span class="r-in"><span>  <span class="va">vcf.report</span> <span class="op">&lt;-</span> <span class="fu">generateVcfReport</span><span class="op">(</span>bam<span class="op">=</span><span class="va">capture.bam</span>, bed<span class="op">=</span><span class="va">capture.bed</span>,</span></span>
<span class="r-in"><span>                                  vcf<span class="op">=</span><span class="va">capture.vcf</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required namespace: VariantAnnotation</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading BED file </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [0.027s]</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading VCF file </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [4.785s]</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Checking BAM file: </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> short-read, paired-end, name-sorted alignment detected</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading paired-end BAM file </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [0.013s]</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Thresholding reads </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [0.001s]</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Extracting base frequences </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [0.105s]</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Oleksii Nikolaienko.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

    </footer></div>





  </body></html>

