<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>generateBedReport — generateBedReport • epialleleR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="generateBedReport — generateBedReport"><meta property="og:description" content="`generateBedReport`, `generateAmpliconReport`, `generateCaptureReport` –
these functions match BAM reads to the set of genomic locations and return
the fraction of reads with an average methylation level passing an arbitrary
threshold."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">epialleleR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.13.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/epialleleR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/values.html">The epialleleR output values &lt;img style="float: right;" src="epialleleR_logo.svg"&gt; </a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/BBCG/epialleleR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>generateBedReport</h1>
    <small class="dont-index">Source: <a href="https://github.com/BBCG/epialleleR/blob/devel/R/generateBedReport.R" class="external-link"><code>R/generateBedReport.R</code></a></small>
    <div class="hidden name"><code>generateBedReport.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>`generateBedReport`, `generateAmpliconReport`, `generateCaptureReport` –
these functions match BAM reads to the set of genomic locations and return
the fraction of reads with an average methylation level passing an arbitrary
threshold.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">generateAmpliconReport</span><span class="op">(</span></span>
<span>  <span class="va">bam</span>,</span>
<span>  <span class="va">bed</span>,</span>
<span>  report.file <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  zero.based.bed <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  match.tolerance <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  threshold.reads <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  threshold.context <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CG"</span>, <span class="st">"CHG"</span>, <span class="st">"CHH"</span>, <span class="st">"CxG"</span>, <span class="st">"CX"</span><span class="op">)</span>,</span>
<span>  min.context.sites <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  min.context.beta <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  max.outofcontext.beta <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  <span class="va">...</span>,</span>
<span>  gzip <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">generateCaptureReport</span><span class="op">(</span></span>
<span>  <span class="va">bam</span>,</span>
<span>  <span class="va">bed</span>,</span>
<span>  report.file <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  zero.based.bed <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  match.min.overlap <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  threshold.reads <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  threshold.context <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CG"</span>, <span class="st">"CHG"</span>, <span class="st">"CHH"</span>, <span class="st">"CxG"</span>, <span class="st">"CX"</span><span class="op">)</span>,</span>
<span>  min.context.sites <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  min.context.beta <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  max.outofcontext.beta <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  <span class="va">...</span>,</span>
<span>  gzip <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">generateBedReport</span><span class="op">(</span></span>
<span>  <span class="va">bam</span>,</span>
<span>  <span class="va">bed</span>,</span>
<span>  report.file <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  zero.based.bed <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  bed.type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"amplicon"</span>, <span class="st">"capture"</span><span class="op">)</span>,</span>
<span>  match.tolerance <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  match.min.overlap <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  threshold.reads <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  threshold.context <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CG"</span>, <span class="st">"CHG"</span>, <span class="st">"CHH"</span>, <span class="st">"CxG"</span>, <span class="st">"CX"</span><span class="op">)</span>,</span>
<span>  min.context.sites <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  min.context.beta <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  max.outofcontext.beta <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  <span class="va">...</span>,</span>
<span>  gzip <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-bam">bam<a class="anchor" aria-label="anchor" href="#arg-bam"></a></dt>
<dd><p>BAM file location string OR preprocessed output of
<code><a href="preprocessBam.html">preprocessBam</a></code> function. Read more about BAM file
requirements and BAM preprocessing at <code><a href="preprocessBam.html">preprocessBam</a></code>.</p></dd>


<dt id="arg-bed">bed<a class="anchor" aria-label="anchor" href="#arg-bed"></a></dt>
<dd><p>Browser Extensible Data (BED) file location string OR object of
class <code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></code> holding genomic coordinates for
regions of interest. The style of seqlevels of BED file/object must be the
same as the style of seqlevels of BAM file/object used. The
BED/<code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></code> rows are <strong>not</strong> sorted
internally. As of now, the strand information is ignored and reads (read
pairs) matching both strands are separately counted and reported.</p></dd>


<dt id="arg-report-file">report.file<a class="anchor" aria-label="anchor" href="#arg-report-file"></a></dt>
<dd><p>file location string to write the BED report. If NULL
(the default) then report is returned as a
<code><a href="https://rdatatable.gitlab.io/data.table/reference/data.table.html" class="external-link">data.table</a></code> object.</p></dd>


<dt id="arg-zero-based-bed">zero.based.bed<a class="anchor" aria-label="anchor" href="#arg-zero-based-bed"></a></dt>
<dd><p>boolean defining if BED coordinates are zero based
(default: FALSE).</p></dd>


<dt id="arg-match-tolerance">match.tolerance<a class="anchor" aria-label="anchor" href="#arg-match-tolerance"></a></dt>
<dd><p>integer for the largest difference between read's and
BED <code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></code> start or end positions during
matching of amplicon-based NGS reads (default: 1).</p></dd>


<dt id="arg-threshold-reads">threshold.reads<a class="anchor" aria-label="anchor" href="#arg-threshold-reads"></a></dt>
<dd><p>boolean defining if sequence reads should be
thresholded before counting reads belonging to variant epialleles (default:
TRUE). Disabling thresholding is possible but makes no sense in the context
of this function, because
all the reads will be assigned to the variant epiallele, which will result
in VEF==1 (in such case `NA` VEF values are returned in order to avoid
confusion). As thresholding is <strong>not</strong> recommended for long-read
sequencing data, this function is <strong>not</strong> recommended for such data
either.</p></dd>


<dt id="arg-threshold-context">threshold.context<a class="anchor" aria-label="anchor" href="#arg-threshold-context"></a></dt>
<dd><p>string defining cytosine methylation context used
for thresholding the reads:</p><ul><li><p>"CG" (the default) – within-the-context: CpG cytosines (called as
  zZ), out-of-context: all the other cytosines (hHxX)</p></li>
<li><p>"CHG" – within-the-context: CHG cytosines (xX), out-of-context: hHzZ</p></li>
<li><p>"CHH" – within-the-context: CHH cytosines (hH), out-of-context: xXzZ</p></li>
<li><p>"CxG" – within-the-context: CG and CHG cytosines (zZxX),
  out-of-context: CHH cytosines (hH)</p></li>
<li><p>"CX" – all cytosines are considered within-the-context, this
  effectively results in no thresholding</p></li>
</ul><p>This option has no effect when read thresholding is disabled.</p></dd>


<dt id="arg-min-context-sites">min.context.sites<a class="anchor" aria-label="anchor" href="#arg-min-context-sites"></a></dt>
<dd><p>non-negative integer for minimum number of cytosines
within the `threshold.context` (default: 2). Reads containing <strong>fewer</strong>
within-the-context cytosines are considered completely unmethylated (thus
belonging to the reference epiallele). This option has no effect when read
thresholding is disabled.</p></dd>


<dt id="arg-min-context-beta">min.context.beta<a class="anchor" aria-label="anchor" href="#arg-min-context-beta"></a></dt>
<dd><p>real number in the range [0;1] (default: 0.5). Reads
with average beta value for within-the-context cytosines <strong>below</strong> this
threshold are considered completely unmethylated (thus belonging to the
reference epiallele). This option has no effect when read thresholding is
disabled.</p></dd>


<dt id="arg-max-outofcontext-beta">max.outofcontext.beta<a class="anchor" aria-label="anchor" href="#arg-max-outofcontext-beta"></a></dt>
<dd><p>real number in the range [0;1] (default: 0.1).
Reads with average beta value for out-of-context cytosines <strong>above</strong>
this threshold are considered completely unmethylated (thus belonging to the
reference epiallele). This option has no effect when read thresholding is
disabled.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>other parameters to pass to the
<code><a href="preprocessBam.html">preprocessBam</a></code> function.
Options have no effect if preprocessed BAM data was supplied as an input.</p></dd>


<dt id="arg-gzip">gzip<a class="anchor" aria-label="anchor" href="#arg-gzip"></a></dt>
<dd><p>boolean to compress the report (default: FALSE).</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>boolean to report progress and timings (default: TRUE).</p></dd>


<dt id="arg-match-min-overlap">match.min.overlap<a class="anchor" aria-label="anchor" href="#arg-match-min-overlap"></a></dt>
<dd><p>integer for the smallest overlap between read's and
BED <code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></code> start or end positions during
matching of capture-based NGS reads (default: 1). If read matches two or more
BED genomic regions, only the first match is taken (input
<code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></code> are <strong>not</strong> sorted internally).</p></dd>


<dt id="arg-bed-type">bed.type<a class="anchor" aria-label="anchor" href="#arg-bed-type"></a></dt>
<dd><p>character string for the type of assay that was used to
produce sequencing reads:</p><ul><li><p>"amplicon" (the default) – used for amplicon-based next-generation
  sequencing when exact coordinates of sequenced fragments are known.
  Matching of reads to genomic ranges are then performed by the read's start
  or end positions, either of which should be no further than
  `match.tolerance` bases away from the start or end position of genomic
  ranges given in BED file/<code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></code> object</p></li>
<li><p>"capture" – used for capture-based next-generation sequencing when
  reads partially overlap with the capture target regions. Read is considered
  to match the genomic range when their overlap is more or equal to
  `match.min.overlap`. If read matches two or more BED genomic regions, only
  the first match is taken (input <code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></code> are
  <strong>not</strong> sorted internally)</p></li>
</ul></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p><code><a href="https://rdatatable.gitlab.io/data.table/reference/data.table.html" class="external-link">data.table</a></code> object containing VEF report for
BED <code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></code> or NULL if report.file was
specified. If BAM file contains reads that would not match to any of BED
<code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></code>, the last row in the report will
contain information on such reads (with seqnames, start and end equal to NA).
The report columns are:</p><ul><li><p>seqnames – reference sequence name</p></li>
<li><p>start – start of genomic region</p></li>
<li><p>end – end of genomic region</p></li>
<li><p>width – width of genomic region</p></li>
<li><p>strand – strand</p></li>
<li><p>... – other columns that were present in BED or metadata columns of
  <code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></code> object</p></li>
<li><p>nreads+ – number of reads (pairs) mapped to the forward ("+") strand</p></li>
<li><p>nreads- – number of reads (pairs) mapped to the reverse ("-") strand</p></li>
<li><p>VEF – frequency of reads passing the threshold</p></li>
</ul></div>
    <div id="details">
    <h2>Details</h2>
    <p>Functions report hypermethylated variant epiallele frequencies (VEF) per
genomic region of interest using BAM and BED files as input. Reads (for
paired-end sequencing alignment files - read pairs as a single
entity) are matched to genomic locations by exact
coordinates (`generateAmpliconReport` or `generateBedReport` with an option
bed.type="amplicon") or minimum overlap (`generateCaptureReport` or
`generateBedReport` with an option bed.type="capture") – the former to be
used for amplicon-based NGS data, while the latter – for the capture-based
NGS data. The function's logic is explained below.</p>
<p>Let's suppose we have a BAM file with four reads, all mapped to the "+"
strand of chromosome 1, positions 1-16. The genomic range is supplied as a
parameter `bed = as("chr1:1-100", "GRanges")`. Assuming the default values
for the thresholding parameters (threshold.reads = TRUE,
threshold.context = "CG", min.context.sites = 2, min.context.beta = 0.5,
max.outofcontext.beta = 0.1), the input and results will look as following:</p>
<table class="table table"><tr><td>methylation string</td><td>threshold</td><td>explained</td></tr><tr><td>...Z..x+.h..x..h.</td><td>below</td><td>min.context.sites &lt; 2 (only one zZ base)</td></tr><tr><td>...Z..z.h..x..h.</td><td>above</td><td>pass all criteria</td></tr><tr><td>...Z..z.h..X..h.</td><td>below</td><td>max.outofcontext.beta &gt; 0.1 (1XH / 3xXhH = 0.33)</td></tr><tr><td>...Z..z.h..z-.h.</td><td>below</td><td>min.context.beta &lt; 0.5 (1Z / 3zZ = 0.33)</td></tr></table><p>Only the second read will satisfy all of the thresholding criteria, leading
to the following BED report (given that all reads map to chr1:+:1-16):</p>
<table class="table table"><tr><td>seqnames</td><td>start</td><td>end</td><td>width</td><td>strand</td><td>nreads+</td><td>nreads-</td><td>VEF</td></tr><tr><td>chr1</td><td>1</td><td>100</td><td>100</td><td>*</td><td>4</td><td>0</td><td>0.25</td></tr></table><p>Please note, that read thresholding by an average methylation level
(as explained above) makes little sense for long-read sequencing alignments,
as such reads can cover multiple regions with very different DNA methylation
properties. Instead, please use <code><a href="extractPatterns.html">extractPatterns</a></code>, limiting
pattern output to the region of interest only.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="preprocessBam.html">preprocessBam</a></code> for preloading BAM data,
<code><a href="generateCytosineReport.html">generateCytosineReport</a></code> for methylation statistics at the level
of individual cytosines, <code><a href="generateVcfReport.html">generateVcfReport</a></code> for evaluating
epiallele-SNV associations, <code><a href="extractPatterns.html">extractPatterns</a></code> for exploring
methylation patterns and <code><a href="plotPatterns.html">plotPatterns</a></code> for pretty plotting
of its output, <code><a href="generateBedEcdf.html">generateBedEcdf</a></code> for analysing the
distribution of per-read beta values, and `epialleleR` vignettes for the
description of usage and sample data.</p>
<p><code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></code> class for working with genomic ranges,
<code><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqlevelsStyle.html" class="external-link">seqlevelsStyle</a></code> function for getting or setting
the seqlevels style.</p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span>  <span class="co"># amplicon data</span></span></span>
<span class="r-in"><span>  <span class="va">amplicon.bam</span>    <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"amplicon010meth.bam"</span>,</span></span>
<span class="r-in"><span>                                 package<span class="op">=</span><span class="st">"epialleleR"</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="va">amplicon.bed</span>    <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"amplicon.bed"</span>,</span></span>
<span class="r-in"><span>                                 package<span class="op">=</span><span class="st">"epialleleR"</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="va">amplicon.report</span> <span class="op">&lt;-</span> <span class="fu">generateAmpliconReport</span><span class="op">(</span>bam<span class="op">=</span><span class="va">amplicon.bam</span>,</span></span>
<span class="r-in"><span>                                            bed<span class="op">=</span><span class="va">amplicon.bed</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading BED file </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [0.007s]</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Checking BAM file: </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> short-read, paired-end, name-sorted alignment detected</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading paired-end BAM file </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [0.005s]</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Thresholding reads </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [0.001s]</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Preparing amplicon report </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [0.013s]</span>
<span class="r-in"><span>  </span></span>
<span class="r-in"><span>  <span class="co"># capture NGS</span></span></span>
<span class="r-in"><span>  <span class="va">capture.bam</span>    <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"capture.bam"</span>,</span></span>
<span class="r-in"><span>                                package<span class="op">=</span><span class="st">"epialleleR"</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="va">capture.bed</span>    <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"capture.bed"</span>,</span></span>
<span class="r-in"><span>                                package<span class="op">=</span><span class="st">"epialleleR"</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="va">capture.report</span> <span class="op">&lt;-</span> <span class="fu">generateCaptureReport</span><span class="op">(</span>bam<span class="op">=</span><span class="va">capture.bam</span>, bed<span class="op">=</span><span class="va">capture.bed</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading BED file </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [0.008s]</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Checking BAM file: </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> short-read, paired-end, name-sorted alignment detected</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading paired-end BAM file </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [0.014s]</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Thresholding reads </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [0.001s]</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Preparing capture report </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [0.016s]</span>
<span class="r-in"><span>  </span></span>
<span class="r-in"><span>  <span class="co"># generateAmpliconReport and generateCaptureReport are just aliases</span></span></span>
<span class="r-in"><span>  <span class="co"># of the generateBedReport</span></span></span>
<span class="r-in"><span>  <span class="va">bed.report</span> <span class="op">&lt;-</span> <span class="fu">generateBedReport</span><span class="op">(</span>bam<span class="op">=</span><span class="va">capture.bam</span>, bed<span class="op">=</span><span class="va">capture.bed</span>,</span></span>
<span class="r-in"><span>                                  bed.type<span class="op">=</span><span class="st">"capture"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading BED file </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [0.019s]</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Checking BAM file: </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> short-read, paired-end, name-sorted alignment detected</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading paired-end BAM file </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [0.014s]</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Thresholding reads </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [0.001s]</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Preparing capture report </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [0.015s]</span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/identical.html" class="external-link">identical</a></span><span class="op">(</span><span class="va">capture.report</span>, <span class="va">bed.report</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] TRUE</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Oleksii Nikolaienko.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

      </footer></div>






  </body></html>

