<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>generateCytosineReport — generateCytosineReport • epialleleR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="generateCytosineReport — generateCytosineReport"><meta name="description" content="This function counts methylated and unmethylated DNA bases taking into the
account average methylation level of the entire sequence read."><meta property="og:description" content="This function counts methylated and unmethylated DNA bases taking into the
account average methylation level of the entire sequence read."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">epialleleR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.13.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/epialleleR.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/values.html">The epialleleR output values &lt;img style="float: right;" src="epialleleR_logo.svg"&gt; </a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/BBCG/epialleleR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>generateCytosineReport</h1>
      <small class="dont-index">Source: <a href="https://github.com/BBCG/epialleleR/blob/devel/R/generateCytosineReport.R" class="external-link"><code>R/generateCytosineReport.R</code></a></small>
      <div class="d-none name"><code>generateCytosineReport.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function counts methylated and unmethylated DNA bases taking into the
account average methylation level of the entire sequence read.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">generateCytosineReport</span><span class="op">(</span></span>
<span>  <span class="va">bam</span>,</span>
<span>  report.file <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  threshold.reads <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  threshold.context <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CG"</span>, <span class="st">"CHG"</span>, <span class="st">"CHH"</span>, <span class="st">"CxG"</span>, <span class="st">"CX"</span><span class="op">)</span>,</span>
<span>  min.context.sites <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  min.context.beta <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  max.outofcontext.beta <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  report.context <span class="op">=</span> <span class="va">threshold.context</span>,</span>
<span>  <span class="va">...</span>,</span>
<span>  gzip <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-bam">bam<a class="anchor" aria-label="anchor" href="#arg-bam"></a></dt>
<dd><p>BAM file location string OR preprocessed output of
<code><a href="preprocessBam.html">preprocessBam</a></code> function. Read more about BAM file
requirements and BAM preprocessing at <code><a href="preprocessBam.html">preprocessBam</a></code>.</p></dd>


<dt id="arg-report-file">report.file<a class="anchor" aria-label="anchor" href="#arg-report-file"></a></dt>
<dd><p>file location string to write the cytosine report. If NULL
(the default) then report is returned as a
<code><a href="https://rdatatable.gitlab.io/data.table/reference/data.table.html" class="external-link">data.table</a></code> object.</p></dd>


<dt id="arg-threshold-reads">threshold.reads<a class="anchor" aria-label="anchor" href="#arg-threshold-reads"></a></dt>
<dd><p>boolean defining if sequence reads (read pairs) should
be thresholded before counting methylated cytosines (default: TRUE).
Disabling thresholding makes the report virtually indistinguishable from the
ones generated by other software, such as Bismark or Illumina DRAGEN Bio IT
Platform. Thresholding is <strong>not</strong> recommended for long-read sequencing
data.</p></dd>


<dt id="arg-threshold-context">threshold.context<a class="anchor" aria-label="anchor" href="#arg-threshold-context"></a></dt>
<dd><p>string defining cytosine methylation context used
for thresholding the reads:</p><ul><li><p>"CG" (the default) — within-the-context: CpG cytosines (called as
  zZ), out-of-context: all the other cytosines (hHxX)</p></li>
<li><p>"CHG" — within-the-context: CHG cytosines (xX), out-of-context: hHzZ</p></li>
<li><p>"CHH" — within-the-context: CHH cytosines (hH), out-of-context: xXzZ</p></li>
<li><p>"CxG" — within-the-context: CG and CHG cytosines (zZxX),
  out-of-context: CHH cytosines (hH)</p></li>
<li><p>"CX" — all cytosines are considered within-the-context, this
  effectively results in no thresholding</p></li>
</ul><p>This option has no effect when read thresholding is disabled.</p></dd>


<dt id="arg-min-context-sites">min.context.sites<a class="anchor" aria-label="anchor" href="#arg-min-context-sites"></a></dt>
<dd><p>non-negative integer for minimum number of cytosines
within the `threshold.context` (default: 2). Reads containing <strong>fewer</strong>
within-the-context cytosines are considered completely unmethylated (all C
are counted as T). This option has no effect when read thresholding is
disabled.</p></dd>


<dt id="arg-min-context-beta">min.context.beta<a class="anchor" aria-label="anchor" href="#arg-min-context-beta"></a></dt>
<dd><p>real number in the range [0;1] (default: 0.5). Reads
with average beta value for within-the-context cytosines <strong>below</strong> this
threshold are considered completely unmethylated (all C are counted as T).
This option has no effect when read thresholding is disabled.</p></dd>


<dt id="arg-max-outofcontext-beta">max.outofcontext.beta<a class="anchor" aria-label="anchor" href="#arg-max-outofcontext-beta"></a></dt>
<dd><p>real number in the range [0;1] (default: 0.1).
Reads with average beta value for out-of-context cytosines <strong>above</strong>
this threshold are considered completely unmethylated (all C are counted as
T). This option has no effect when read thresholding is disabled.</p></dd>


<dt id="arg-report-context">report.context<a class="anchor" aria-label="anchor" href="#arg-report-context"></a></dt>
<dd><p>string defining cytosine methylation context to report
(default: value of `threshold.context`).</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>other parameters to pass to the
<code><a href="preprocessBam.html">preprocessBam</a></code> function.
Options have no effect if preprocessed BAM data was supplied as an input.</p></dd>


<dt id="arg-gzip">gzip<a class="anchor" aria-label="anchor" href="#arg-gzip"></a></dt>
<dd><p>boolean to compress the report (default: FALSE).</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>boolean to report progress and timings (default: TRUE).</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p><code><a href="https://rdatatable.gitlab.io/data.table/reference/data.table.html" class="external-link">data.table</a></code> object containing cytosine
report in Bismark-like format or NULL if report.file was specified. The
report columns are:</p><ul><li><p>rname — reference sequence name (as in BAM)</p></li>
<li><p>strand — strand</p></li>
<li><p>pos — cytosine position</p></li>
<li><p>context — methylation context</p></li>
<li><p>meth — number of methylated cytosines</p></li>
<li><p>unmeth — number of unmethylated cytosines</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The function reports cytosine methylation information using BAM file or data
as an input. In contrast to the other currently available software, reads
(for paired-end sequencing alignment files - read
pairs as a single entity) can be thresholded by their average
methylation level before counting methylated bases, effectively resulting in
hypermethylated variant epiallele frequency (VEF) being reported instead of
beta value. The function's logic is explained below.</p>
<p>Let's suppose we have a BAM file with four reads, all mapped to the "+"
strand of chromosome 1, positions 1-16. Assuming the default values
for the thresholding parameters (threshold.reads = TRUE,
threshold.context = "CG", min.context.sites = 2, min.context.beta = 0.5,
max.outofcontext.beta = 0.1), the input and results will look as following:</p>
<table class="table table"><tr><td>methylation string</td><td>threshold</td><td>explained</td><td>methylation reported</td></tr><tr><td>...Z..x+.h..x..h.</td><td>below</td><td>min.context.sites &lt; 2 (only one zZ base)</td><td>all cytosines unmethylated</td></tr><tr><td>...Z..z.h..x..h.</td><td>above</td><td>pass all criteria</td><td>only C4 (Z at position 4) is methylated</td></tr><tr><td>...Z..z.h..X..h.</td><td>below</td><td>max.outofcontext.beta &gt; 0.1 (1XH / 3xXhH = 0.33)</td><td>all cytosines unmethylated</td></tr><tr><td>...Z..z.h..z-.h.</td><td>below</td><td>min.context.beta &lt; 0.5 (1Z / 3zZ = 0.33)</td><td>all cytosines unmethylated</td></tr></table><p>Only the second read will satisfy all of the thresholding criteria, leading
to the following CX report (given that all reads map to chr1:+:1-16):</p>
<table class="table table"><tr><td>rname</td><td>strand</td><td>pos</td><td>context</td><td>meth</td><td>unmeth</td></tr><tr><td>chr1</td><td>+</td><td>4</td><td>CG</td><td>1</td><td>3</td></tr><tr><td>chr1</td><td>+</td><td>7</td><td>CG</td><td>0</td><td>3</td></tr><tr><td>chr1</td><td>+</td><td>9</td><td>CHH</td><td>0</td><td>4</td></tr><tr><td>chr1</td><td>+</td><td>12</td><td>CHG</td><td>0</td><td>3</td></tr><tr><td>chr1</td><td>+</td><td>15</td><td>CHH</td><td>0</td><td>4</td></tr></table><p>With the thresholding disabled (threshold.reads = FALSE) all methylated bases
will retain their status, so the CX report will be similar to the reports
produced by other methylation callers (such as Bismark or Illumina DRAGEN Bio
IT Platform):</p>
<table class="table table"><tr><td>rname</td><td>strand</td><td>pos</td><td>context</td><td>meth</td><td>unmeth</td></tr><tr><td>chr1</td><td>+</td><td>4</td><td>CG</td><td>4</td><td>0</td></tr><tr><td>chr1</td><td>+</td><td>7</td><td>CG</td><td>0</td><td>3</td></tr><tr><td>chr1</td><td>+</td><td>9</td><td>CHH</td><td>0</td><td>4</td></tr><tr><td>chr1</td><td>+</td><td>12</td><td>CHG</td><td>1</td><td>2</td></tr><tr><td>chr1</td><td>+</td><td>15</td><td>CHH</td><td>0</td><td>4</td></tr></table><p>Other notes:</p>
<p>To produce conventional cytosine reports without thresholding by
within-context methylation level though
minimally affected by incomplete cytosine conversion, run this method with
the following parameters: `threshold.reads=TRUE`, `threshold.context="CG"`,
`min.context.sites=0`, `min.context.beta=0`, `max.outofcontext.beta=0.1`.
All cytosines within reads (read pairs) having more than 10<!-- % out-of-context -->
cytosines methylated, will be effectively treated as unmethylated ones.</p>
<p>Methylation string bases in unknown context ("uU") are simply ignored, which,
to the best of our knowledge, is consistent with the behaviour of other
tools.</p>
<p>In order to mitigate the effect of sequencing errors (leading to rare
variations in the methylation context, as in reads 1 and 4 above), the
context present in more than 50% of the reads is assumed to be correct,
while all bases at the same position but having other methylation context
are simply ignored. This allows reports to be prepared without using the
reference genome sequence.</p>
<p>The downside of not using the reference genome sequence is the inability to
determine the actual sequence of triplet for every base in the cytosine
report. Therefore this sequence is not reported, and this won't change
until such information will be considered as worth adding.</p>
<p>Please also note, that read thresholding by an average methylation level
(as explained above) makes little sense for long-read sequencing alignments,
as such reads can cover multiple regions with very different DNA methylation
properties.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p>`values` vignette for a comparison and visualisation of epialleleR
output values for various input files. `epialleleR` vignette for the
description of usage and sample data.</p>
<p><code><a href="preprocessBam.html">preprocessBam</a></code> for preloading BAM data,
<code><a href="generateBedReport.html">generateBedReport</a></code> for genomic region-based statistics,
<code><a href="generateVcfReport.html">generateVcfReport</a></code> for evaluating epiallele-SNV associations,
<code><a href="extractPatterns.html">extractPatterns</a></code> for exploring methylation patterns and
<code><a href="plotPatterns.html">plotPatterns</a></code> for pretty plotting of its output,
<code><a href="generateBedEcdf.html">generateBedEcdf</a></code> for analysing the distribution of per-read
beta values.</p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span>  <span class="va">capture.bam</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"capture.bam"</span>, package<span class="op">=</span><span class="st">"epialleleR"</span><span class="op">)</span></span></span>
<span class="r-in"><span>  </span></span>
<span class="r-in"><span>  <span class="co"># CpG report with thresholding</span></span></span>
<span class="r-in"><span>  <span class="va">cg.report</span> <span class="op">&lt;-</span> <span class="fu">generateCytosineReport</span><span class="op">(</span><span class="va">capture.bam</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Checking BAM file: </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> short-read, paired-end, name-sorted alignment detected</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading paired-end BAM file </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [0.012s]</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Thresholding reads </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [0.001s]</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Preparing cytosine report </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [0.013s]</span>
<span class="r-in"><span>  </span></span>
<span class="r-in"><span>  <span class="co"># CX report without thresholding</span></span></span>
<span class="r-in"><span>  <span class="va">cx.report</span> <span class="op">&lt;-</span> <span class="fu">generateCytosineReport</span><span class="op">(</span><span class="va">capture.bam</span>, threshold.reads<span class="op">=</span><span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span>               report.context<span class="op">=</span><span class="st">"CX"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Checking BAM file: </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> short-read, paired-end, name-sorted alignment detected</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading paired-end BAM file </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [0.011s]</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Preparing cytosine report </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [0.017s]</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Oleksii Nikolaienko.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

    </footer></div>





  </body></html>

