<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>plotPatterns — plotPatterns • epialleleR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="plotPatterns — plotPatterns"><meta property="og:description" content="This convenience function plots methylation patterns (epialleles) previously
extracted by extractPatterns."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">epialleleR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.13.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/epialleleR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/values.html">The epialleleR output values &lt;img style="float: right;" src="epialleleR_logo.svg"&gt; </a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/BBCG/epialleleR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>plotPatterns</h1>
    <small class="dont-index">Source: <a href="https://github.com/BBCG/epialleleR/blob/devel/R/plotPatterns.R" class="external-link"><code>R/plotPatterns.R</code></a></small>
    <div class="hidden name"><code>plotPatterns.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This convenience function plots methylation patterns (epialleles) previously
extracted by <code><a href="extractPatterns.html">extractPatterns</a></code>.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plotPatterns</span><span class="op">(</span></span>
<span>  <span class="va">patterns</span>,</span>
<span>  order.by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"beta"</span>, <span class="st">"count"</span><span class="op">)</span>,</span>
<span>  beta.range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>  bin.context <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CG"</span>, <span class="st">"CHG"</span>, <span class="st">"CHH"</span>, <span class="st">"CxG"</span>, <span class="st">"CX"</span><span class="op">)</span>,</span>
<span>  nbins <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  npatterns.per.bin <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  plot.context <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CG"</span>, <span class="st">"CHG"</span>, <span class="st">"CHH"</span>, <span class="st">"CxG"</span>, <span class="st">"CX"</span><span class="op">)</span>,</span>
<span>  genomic.scale <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"continuous"</span>, <span class="st">"discrete"</span><span class="op">)</span>,</span>
<span>  breaks <span class="op">=</span> <span class="st">"auto"</span>,</span>
<span>  marginal <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"density"</span>, <span class="st">"count"</span><span class="op">)</span>,</span>
<span>  marginal.position <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"left"</span>, <span class="st">"right"</span><span class="op">)</span>,</span>
<span>  marginal.transform <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"identity"</span>, <span class="st">"log10"</span><span class="op">)</span>,</span>
<span>  marginal.limits <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  marginal.size <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>  <span class="va">...</span>,</span>
<span>  tag <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"none"</span>, <span class="st">"count"</span>, <span class="st">"beta"</span>, <span class="st">"pattern"</span><span class="op">)</span>,</span>
<span>  tag.size <span class="op">=</span> <span class="fl">2.5</span>,</span>
<span>  tag.colour <span class="op">=</span> <span class="st">"#87654c"</span>,</span>
<span>  tag.fill <span class="op">=</span> <span class="st">"lemonchiffon"</span>,</span>
<span>  title <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  subtitle <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  context.size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>  base.size <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  methylation.fill <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"grey97"</span>, <span class="st">"grey10"</span><span class="op">)</span>,</span>
<span>  plot <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-patterns">patterns<a class="anchor" aria-label="anchor" href="#arg-patterns"></a></dt>
<dd><p>output of <code><a href="preprocessBam.html">preprocessBam</a></code> function
(methylation patterns as a <code><a href="https://rdatatable.gitlab.io/data.table/reference/data.table.html" class="external-link">data.table</a></code> object).</p></dd>


<dt id="arg-order-by">order.by<a class="anchor" aria-label="anchor" href="#arg-order-by"></a></dt>
<dd><p>string defining order of patterns on the plot (default order
by: "beta").</p></dd>


<dt id="arg-beta-range">beta.range<a class="anchor" aria-label="anchor" href="#arg-beta-range"></a></dt>
<dd><p>numeric vector of length 2 for the range of average
pattern beta values represented on the plot (default: [0;1]).</p></dd>


<dt id="arg-bin-context">bin.context<a class="anchor" aria-label="anchor" href="#arg-bin-context"></a></dt>
<dd><p>string defining cytosine methylation context used
to calculate average beta value of a pattern that is further used to assign
patterns to bins:</p><ul><li><p>"CG" (the default) – CpG cytosines (called as zZ)</p></li>
<li><p>"CHG" – CHG cytosines (xX)</p></li>
<li><p>"CHH" – CHH cytosines (hH)</p></li>
<li><p>"CxG" – CG and CHG cytosines (zZxX)</p></li>
<li><p>"CX" – all cytosines</p></li>
</ul></dd>


<dt id="arg-nbins">nbins<a class="anchor" aria-label="anchor" href="#arg-nbins"></a></dt>
<dd><p>a single integer defining the number of bins (i.e., intervals
within `beta.range`). Default: 10.</p></dd>


<dt id="arg-npatterns-per-bin">npatterns.per.bin<a class="anchor" aria-label="anchor" href="#arg-npatterns-per-bin"></a></dt>
<dd><p>integer vector for the number of the most abundant
patterns selected from each bin (default: 2). When of length 1, the same
number of patterns will be taken. When of length `nbins`, allows to
fine-tune the number of selected patterns from each bin. Setting to `Inf`
effectively results in plotting all patterns.</p></dd>


<dt id="arg-plot-context">plot.context<a class="anchor" aria-label="anchor" href="#arg-plot-context"></a></dt>
<dd><p>string defining methylation context of cytosines
included in the plot (default: "CG"; for the range of available values, see
`bin.context` above).</p></dd>


<dt id="arg-genomic-scale">genomic.scale<a class="anchor" aria-label="anchor" href="#arg-genomic-scale"></a></dt>
<dd><p>string for the type of genomic position scale
of the plot: either "continuous" (the default) or "discrete".</p></dd>


<dt id="arg-breaks">breaks<a class="anchor" aria-label="anchor" href="#arg-breaks"></a></dt>
<dd><p>a vector of breaks for the genomic position scale of the plot.
If "auto" (the default), breaks for continuous scale are computed by the
default ggplot2 routines, while breaks for discrete scale are a subset of
plotted positions selected using <code><a href="https://rdrr.io/r/base/pretty.html" class="external-link">pretty</a></code>.
Possible values: <code><a href="https://ggplot2.tidyverse.org/reference/waiver.html" class="external-link">ggplot2::waiver()</a></code> for ggplot2
defaults, integer vector of breaks for continuous scale, or character vector
of breaks for discrete scale.</p></dd>


<dt id="arg-marginal">marginal<a class="anchor" aria-label="anchor" href="#arg-marginal"></a></dt>
<dd><p>string for the type of marginal plot: either "density"
(probability density of average beta values of all patterns; the default)
or "count" (counts of plotted patterns). "none" is not implemented yet;
create an issue if interested.</p></dd>


<dt id="arg-marginal-position">marginal.position<a class="anchor" aria-label="anchor" href="#arg-marginal-position"></a></dt>
<dd><p>string for the position of marginal plot: either
"left" (the default) or "right" (not implemented yet; create an issue if
interested).</p></dd>


<dt id="arg-marginal-transform">marginal.transform<a class="anchor" aria-label="anchor" href="#arg-marginal-transform"></a></dt>
<dd><p>string for the transformation of marginal scale
(default: "identity"). Check
<code><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">ggplot2::scale_x_continuous()</a></code> for
more details.</p></dd>


<dt id="arg-marginal-limits">marginal.limits<a class="anchor" aria-label="anchor" href="#arg-marginal-limits"></a></dt>
<dd><p>limits of marginal scale (default: NULL). Check
<code><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">ggplot2::scale_x_continuous()</a></code> for
more details.</p></dd>


<dt id="arg-marginal-size">marginal.size<a class="anchor" aria-label="anchor" href="#arg-marginal-size"></a></dt>
<dd><p>numeric in range (0;1) for the relative width of the
marginal plot (default: 0.25).</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>additional arguments passed to
<code><a href="https://rdrr.io/r/stats/density.html" class="external-link">stats::density()</a></code> call used in marginal density
plot. Possible value: <code>adjust=0.25</code>.</p></dd>


<dt id="arg-tag">tag<a class="anchor" aria-label="anchor" href="#arg-tag"></a></dt>
<dd><p>string for optional tagging of patterns with their count
("count"), average beta value ("beta"), or pattern ID ("pattern").
Default: "none".</p></dd>


<dt id="arg-tag-size">tag.size<a class="anchor" aria-label="anchor" href="#arg-tag-size"></a></dt>
<dd><p>numeric for the font size of the tag text
(in millimetres; default: 2.5).</p></dd>


<dt id="arg-tag-colour">tag.colour<a class="anchor" aria-label="anchor" href="#arg-tag-colour"></a></dt>
<dd><p>string for the colour of of the tag text.
Default: "#87654c".</p></dd>


<dt id="arg-tag-fill">tag.fill<a class="anchor" aria-label="anchor" href="#arg-tag-fill"></a></dt>
<dd><p>string for the colour of of the tag background.
Default: "lemonchiffon".</p></dd>


<dt id="arg-title">title<a class="anchor" aria-label="anchor" href="#arg-title"></a></dt>
<dd><p>the title of the plot. When `TRUE` (the default), a genomic
region from which patterns were extracted. Other possible values: anything
that can be converted to string, or `NULL` for no title.</p></dd>


<dt id="arg-subtitle">subtitle<a class="anchor" aria-label="anchor" href="#arg-subtitle"></a></dt>
<dd><p>the subtitle of the plot. When `TRUE` (the default), a
number of patterns plotted. Other possible values: anything
that can be converted to string, or `NULL` for no subtitle.</p></dd>


<dt id="arg-context-size">context.size<a class="anchor" aria-label="anchor" href="#arg-context-size"></a></dt>
<dd><p>a numeric vector with sizes of circles representing
cytosines within each of three contexts: CHH, CHG, and CG
(default: c(1, 2, 3)).</p></dd>


<dt id="arg-base-size">base.size<a class="anchor" aria-label="anchor" href="#arg-base-size"></a></dt>
<dd><p>numeric for the font size of the text for highlighted bases
(in millimetres; default: 3).</p></dd>


<dt id="arg-methylation-fill">methylation.fill<a class="anchor" aria-label="anchor" href="#arg-methylation-fill"></a></dt>
<dd><p>a vector of length 2 for colours representing
unmethylated and methylated cytosines, respectively. These colours are also
mapped to the lowest (0) and highest (1) possible beta values
to represent average beta values of methylation patterns and create a
gradient fill of a marginal density plot. Default: c("grey97", "grey10").</p></dd>


<dt id="arg-plot">plot<a class="anchor" aria-label="anchor" href="#arg-plot"></a></dt>
<dd><p>boolean. If `TRUE` (the default), patterns are plotted, and the
selected ones are silently returned as a <code><a href="https://rdatatable.gitlab.io/data.table/reference/data.table.html" class="external-link">data.table</a></code>
object. If `FALSE`, the <code><a href="https://gtable.r-lib.org/reference/gtable.html" class="external-link">grob table</a></code> object
is returned instead.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>boolean to report basic info on input and output.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>the plot and (silently) the <code><a href="https://rdatatable.gitlab.io/data.table/reference/data.table.html" class="external-link">data.table</a></code>
object containing plotted methylation patterns (if `plot==TRUE`),
or <code><a href="https://gtable.r-lib.org/reference/gtable.html" class="external-link">grob table</a></code> object (if `plot==FALSE`).</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>As the number of methylation patterns can be quite large, by default, the
function plots <strong>the most abundant unique patterns</strong> only. The complete
logic is as follows:</p><ul><li><p>from the input methylation patterns, all unique patterns are
  extracted and counted</p></li>
<li><p>unique patterns are split in bins by their average beta value</p></li>
<li><p>most abundant unique methylation patterns from each bin are plotted
  and silently returned</p></li>
</ul><p>On the resulting plot, each cytosine is shown as a circle,
where the size of that circle represents cytosine context
and the fill encodes methylation status. If available, highlighted bases are
shown as labels of different colours.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="extractPatterns.html">extractPatterns</a></code> for extracting methylation patterns,
<code><a href="preprocessBam.html">preprocessBam</a></code> for preloading BAM data,
<code><a href="generateCytosineReport.html">generateCytosineReport</a></code> for methylation statistics at the level
of individual cytosines, <code><a href="generateBedReport.html">generateBedReport</a></code> for genomic
region-based statistics, <code><a href="generateVcfReport.html">generateVcfReport</a></code> for evaluating
epiallele-SNV associations, <code><a href="generateBedEcdf.html">generateBedEcdf</a></code> for analysing the
distribution of per-read beta values, and `epialleleR` vignettes for the
description of usage and sample data.</p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span>  <span class="co"># amplicon data</span></span></span>
<span class="r-in"><span>  <span class="va">amplicon.bam</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"amplicon010meth.bam"</span>,</span></span>
<span class="r-in"><span>                              package<span class="op">=</span><span class="st">"epialleleR"</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="va">custom.range</span> <span class="op">&lt;-</span> <span class="fu">as</span><span class="op">(</span><span class="st">"chr17:43124861-43125150"</span>, <span class="st">"GRanges"</span><span class="op">)</span></span></span>
<span class="r-in"><span>  </span></span>
<span class="r-in"><span>  <span class="co"># let's get our patterns</span></span></span>
<span class="r-in"><span>  <span class="va">patterns</span> <span class="op">&lt;-</span> <span class="fu"><a href="extractPatterns.html">extractPatterns</a></span><span class="op">(</span>bam<span class="op">=</span><span class="va">amplicon.bam</span>, bed<span class="op">=</span><span class="va">custom.range</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Checking BAM file: </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> short-read, paired-end, name-sorted alignment detected</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading paired-end BAM file </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [0.004s]</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Extracting methylation patterns </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [0.011s]</span>
<span class="r-in"><span>  </span></span>
<span class="r-in"><span>  <span class="co"># default plot + silently returned plotted patterns</span></span></span>
<span class="r-in"><span>  <span class="va">selected.patterns</span> <span class="op">&lt;-</span> <span class="fu">plotPatterns</span><span class="op">(</span><span class="va">patterns</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 84 patterns supplied</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 21 unique</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 7 most frequent unique patterns were selected for plotting using 10 beta value bins:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [0,0.1) [0.1,0.2) [0.2,0.3) [0.3,0.4) [0.4,0.5) [0.5,0.6) [0.6,0.7) [0.7,0.8) [0.8,0.9) [0.9,1]</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>       2         1         1         0         0         0         0         1         0       2</span>
<span class="r-plt img"><img src="plotPatterns-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span>  </span></span>
<span class="r-in"><span>  <span class="co"># all unique patterns with their counts as a margin, categorical positions,</span></span></span>
<span class="r-in"><span>  <span class="co"># tagged with pattern IDs, returned as a `gtable` object</span></span></span>
<span class="r-in"><span>  <span class="va">tbl</span> <span class="op">&lt;-</span> <span class="fu">plotPatterns</span><span class="op">(</span><span class="va">patterns</span>, npatterns.per.bin<span class="op">=</span><span class="cn">Inf</span>, marginal<span class="op">=</span><span class="st">"count"</span>,</span></span>
<span class="r-in"><span>                      genomic.scale<span class="op">=</span><span class="st">"discrete"</span>, tag<span class="op">=</span><span class="st">"pattern"</span>, plot<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 84 patterns supplied</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 21 unique</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 21 most frequent unique patterns were selected for plotting using 10 beta value bins:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [0,0.1) [0.1,0.2) [0.2,0.3) [0.3,0.4) [0.4,0.5) [0.5,0.6) [0.6,0.7) [0.7,0.8) [0.8,0.9) [0.9,1]</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>      14         1         1         0         0         0         0         1         0       4</span>
<span class="r-in"><span>  </span></span>
<span class="r-in"><span>  <span class="co"># which can be plotted later</span></span></span>
<span class="r-in"><span>  <span class="fu">grid</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grid/grid.newpage.html" class="external-link">grid.newpage</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="fu">grid</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grid/grid.draw.html" class="external-link">grid.draw</a></span><span class="op">(</span><span class="va">tbl</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="plotPatterns-2.png" alt="" width="700" height="433"></span>
<span class="r-in"><span>  </span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Oleksii Nikolaienko.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

      </footer></div>






  </body></html>

