<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>preprocessBam — preprocessBam • epialleleR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="preprocessBam — preprocessBam"><meta property="og:description" content="This function reads and preprocesses BAM file."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">epialleleR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.9.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/epialleleR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/BBCG/epialleleR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>preprocessBam</h1>
    <small class="dont-index">Source: <a href="https://github.com/BBCG/epialleleR/blob/HEAD/R/preprocessBam.R" class="external-link"><code>R/preprocessBam.R</code></a></small>
    <div class="hidden name"><code>preprocessBam.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function reads and preprocesses BAM file.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">preprocessBam</span><span class="op">(</span></span>
<span>  <span class="va">bam.file</span>,</span>
<span>  min.mapq <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  min.baseq <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  skip.duplicates <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  nthreads <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>bam.file</dt>
<dd><p>BAM file location string.</p></dd>


<dt>min.mapq</dt>
<dd><p>non-negative integer threshold for minimum read mapping
quality (default: 0).</p></dd>


<dt>min.baseq</dt>
<dd><p>non-negative integer threshold for minimum nucleotide base
quality (default: 0).</p></dd>


<dt>skip.duplicates</dt>
<dd><p>boolean defining if duplicate aligned reads should be
skipped (default: FALSE). Option has no effect if duplicate reads were not
marked by alignment software.</p></dd>


<dt>nthreads</dt>
<dd><p>non-negative integer for the number of additional HTSlib
threads to be used during BAM file decompression (default: 1). Two threads
(and usually no more than two) make sense for the files larger than 100 MB.</p></dd>


<dt>verbose</dt>
<dd><p>boolean to report progress and timings (default: TRUE).</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p><code><a href="https://Rdatatable.gitlab.io/data.table/reference/data.table.html" class="external-link">data.table</a></code> object containing preprocessed
BAM data.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>The function loads and preprocesses BAM file, saving time when multiple
analyses are to be performed on large input files. Currently, HTSlib
is used to read the data, therefore it is possible to speed up the loading
by means of HTSlib threads.</p>
<p>This function is also called internally when BAM file location is supplied as
an input for other `epialleleR` methods.</p>
<p>`preprocessBam` currently accepts only BAM files that are derived from
paired-end sequencing (create an issue if you need to process single-end BAM
files). During preprocessing, paired reads are merged according to their base 
quality: nucleotide base with the highest value in the QUAL string is taken,
unless its quality is less than `min.baseq`, which results in no information
for that particular position ("-"/"N"). These merged reads are then
processed as a single entity in all `epialleleR` methods. Due to merging,
overlapping bases in read pairs are counted only once, and the base with the
highest quality is taken.</p>
<p>It is also a requirement currently that BAM file is sorted by QNAME instead
of genomic location (i.e., "unsorted") to perform merging of paired-end
reads. Error message is shown if it is sorted by genomic location, in this
case please sort it by QNAME using 'samtools sort -n -o out.bam in.bam'.</p>
<p>Please also note that for all its methods, `epialleleR` requires genomic
strand (XG tag) and a methylation call string (XM tag) to be present in a
BAM file - i.e., methylation calling must be
performed after read mapping/alignment by your software of choice.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="generateCytosineReport.html">generateCytosineReport</a></code> for methylation statistics at
the level of individual cytosines, <code><a href="generateBedReport.html">generateBedReport</a></code> for
genomic region-based statistics, <code><a href="generateVcfReport.html">generateVcfReport</a></code> for
evaluating epiallele-SNV associations, <code><a href="extractPatterns.html">extractPatterns</a></code> for
exploring methylation patterns, <code><a href="generateBedEcdf.html">generateBedEcdf</a></code> for
analysing the distribution of per-read beta values, and `epialleleR`
vignettes for the description of usage and sample data.</p>
<p>Sequence Alignment/Map <a href="https://samtools.github.io/hts-specs/SAMv1.pdf" class="external-link">format specifications</a>,
duplicate alignments marking by <a href="http://www.htslib.org/doc/samtools-markdup.html" class="external-link">Samtools</a>
and <a href="https://support.illumina.com/content/dam/illumina-support/help/Illumina_DRAGEN_Bio_IT_Platform_v3_7_1000000141465/Content/SW/Informatics/Dragen/DuplicateMarking_fDG.htm" class="external-link">Illumina DRAGEN Bio IT Platform</a>.</p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span>  <span class="va">capture.bam</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"capture.bam"</span>, package<span class="op">=</span><span class="st">"epialleleR"</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="va">bam.data</span>    <span class="op">&lt;-</span> <span class="fu">preprocessBam</span><span class="op">(</span><span class="va">capture.bam</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading BAM file</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  [0.017s]</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Oleksii Nikolaienko.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

