<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>The epialleleR User's Guide • epialleleR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="The epialleleR User's Guide">
<meta property="og:description" content="epialleleR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">epialleleR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.11</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/epialleleR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/BBCG/epialleleR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>The epialleleR User’s Guide</h1>
            
            <h4 data-toc-skip class="date">05 October, 2021</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/BBCG/epialleleR/blob/master/vignettes/epialleleR.Rmd" class="external-link"><code>vignettes/epialleleR.Rmd</code></a></small>
      <div class="hidden name"><code>epialleleR.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>A comprehensive guide to using the epialleleR package for calling the hypermethylated epiallele frequencies from next-generation sequencing data</p>
    </div>
    
<hr>
<div align="right">
<svg xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:cc="http://creativecommons.org/ns#" xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#" xmlns:svg="http://www.w3.org/2000/svg" xmlns="http://www.w3.org/2000/svg" xmlns:sodipodi="http://sodipodi.sourceforge.net/DTD/sodipodi-0.dtd" xmlns:inkscape="http://www.inkscape.org/namespaces/inkscape" width="73.043045mm" height="26.458332mm" viewbox="0 0 73.043044 26.458332" version="1.1" id="svg756" sodipodi:docname="logo.svg" inkscape:version="1.0.2 (e86c8708, 2021-01-15)"><namedview pagecolor="#ffffff" bordercolor="#666666" borderopacity="1" objecttolerance="10" gridtolerance="10" guidetolerance="10" inkscape:pageopacity="0" inkscape:pageshadow="2" inkscape:window-width="1402" inkscape:window-height="1226" id="namedview21" showgrid="false" inkscape:zoom="1.514647" inkscape:cx="187.67324" inkscape:cy="28.002964" inkscape:window-x="977" inkscape:window-y="126" inkscape:window-maximized="0" inkscape:current-layer="svg756" inkscape:document-rotation="0"></namedview><defs id="defs750"></defs><metadata id="metadata753"><rdf><work rdf:about=""><format>image/svg+xml</format><type rdf:resource="http://purl.org/dc/dcmitype/StillImage"></type><title></title></work></rdf></metadata><path style="fill:#5fbcd3;stroke:none;stroke-width:0.289798px;stroke-linecap:butt;stroke-linejoin:miter;stroke-opacity:1" d="M 51.943437,5.0581871 H 70.356302 C 73.52237,5.0774447 73.012833,5.586231 73.012833,2.5290839 73.012833,1.2613548 72.020815,0 70.356302,0 H 51.943437 c -1.664513,0 -2.660729,1.2613548 -2.656511,2.5290839 0.0039,1.2613356 1.000375,2.5291032 2.656511,2.5291032 z" id="path29078"></path><path style="fill:#5fbcd3;stroke:none;stroke-width:0.289798px;stroke-linecap:butt;stroke-linejoin:miter;stroke-opacity:1" d="m 51.943437,12.191575 h 18.412865 c 1.656155,0 2.656531,-1.267748 2.656531,-2.5291029 0,-3.0571856 0.509537,-2.5476675 -2.656531,-2.5290839 H 51.943437 c -1.664513,0 -2.660729,1.2613548 -2.656511,2.5290839 0.0039,1.2613359 1.000375,2.5291029 2.656511,2.5291029 z" id="path29080"></path><path style="fill:#5fbcd3;stroke:none;stroke-width:0.289798px;stroke-linecap:butt;stroke-linejoin:miter;stroke-opacity:1" d="m 51.943437,19.324963 h 10.260479 c 1.656156,0 2.652352,-1.267748 2.656531,-2.529103 0.0039,-1.267729 -0.992018,-2.529084 -2.656531,-2.529084 H 51.943437 c -1.664513,0 -2.660729,1.261355 -2.656511,2.529084 0.0039,1.261336 1.000375,2.529103 2.656511,2.529103 z" id="path29082"></path><text xml:space="preserve" style="font-size:10.8698px;line-height:1.25;font-family:sans-serif;letter-spacing:0.509524px;word-spacing:0px;stroke-width:0.509524" x="143.61124" y="19.867887" id="text29086"><tspan id="tspan29084" style="text-align:end;letter-spacing:0.509524px;text-anchor:end;fill:#0088aa;stroke-width:0.509524" x="48.224136" y="19.867887" dy="0" rotate="0 0 0 0 0 0 0 0 0 0">epiallele</tspan></text><path style="fill:#5fbcd3;stroke:none;stroke-width:0.289798px;stroke-linecap:butt;stroke-linejoin:miter;stroke-opacity:1" d="m 51.943437,26.458332 h 5.165221 c 1.656156,0 2.652352,-1.267748 2.656531,-2.529103 0.0039,-1.267729 -0.992018,-2.529084 -2.656531,-2.529084 h -5.165221 c -1.664513,0 -2.660729,1.261355 -2.656511,2.529084 0.0039,1.261336 1.000375,2.529103 2.656511,2.529103 z" id="path29088"></path><path style="fill:#5fbcd3;stroke:none;stroke-width:0.289798px;stroke-linecap:butt;stroke-linejoin:miter;stroke-opacity:1" d="m 65.191081,26.458332 h 5.165221 c 1.656155,0 2.652352,-1.267748 2.656531,-2.529103 0.0039,-1.267729 -0.992018,-2.529084 -2.656531,-2.529084 h -5.165221 c -1.664513,0 -2.660729,1.261355 -2.656511,2.529084 0.0039,1.261336 1.000375,2.529103 2.656511,2.529103 z" id="path29090"></path></svg>
</div>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor" aria-hidden="true"></a>Introduction</h1>
<p>The cytosine DNA methylation is an important epigenetic mechanism for regulation of gene expression. Abnormal methylation is linked to several diseases, being for example the most common molecular lesion in cancer cell<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>. Multiple studies suggest that alterations in DNA methylation, despite occurring at a low mosaic level, may confer increased risk of cancer later in life<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a>.</p>
<p>The cytosine methylation levels within relatively small regions of the human genome are thought to be often concordant, resulting in a limited number of distinct methylation patterns of short sequencing reads<a href="#fn3" class="footnote-ref" id="fnref3"><sup>3</sup></a>. Due to the cell-to-cell variations in methylation, DNA purified from tissue samples contains a mix of hyper- and hypomethylated alleles with varying ratios that depend on the genomic region and tissue type.</p>
<p>Unsurprisingly, when the frequencies of hypermethylated epialleles are low (e.g. 1e-02 and lower) and cytosine methylation levels are averaged and reported using conventional algorithms, the identification of such hypermethylated epialleles becomes nearly impossible. In order to increase the sensitivity of DNA methylation analysis we have developed <em><code>epialleleR</code></em> — an R package for calling hypermethylated variant epiallele frequencies (VEF).</p>
<p><em><code>epialleleR</code></em> is a very fast and scalable solution for analysis of data obtained by next-generation sequencing of bisulfite-treated DNA samples. The minimum requirement for the input is a Binary Alignment Map (BAM) file containing sequencing reads. These reads can be obtained from either deep or ultra-deep sequencing, using either narrowly targeted gene panels (amplicon sequencing), larger methylation capture panels, or even whole-genome approaches.</p>
<div id="current-features" class="section level2">
<h2 class="hasAnchor">
<a href="#current-features" class="anchor" aria-hidden="true"></a>Current Features</h2>
<ul>
<li>In addition to conventional reporting of cytosine DNA methylation levels (beta values), <em><code>epialleleR</code></em> can call variant epiallele frequencies (VEF) of hypermethylated alleles at the level of individual cytosines (<em><code>generateCytosineReport</code></em>) or supplied genomic regions (<em><code>generateBedReport</code></em>)</li>
<li>Potential bimodality of methylation for genomic regions of interest can be assessed using <em><code>generateBedEcdf</code></em> method</li>
<li>The association of single-nucleotide variations with the epiallelic status can be tested using <em><code>generateVcfReport</code></em>
</li>
</ul>
</div>
<div id="processing-speed" class="section level2">
<h2 class="hasAnchor">
<a href="#processing-speed" class="anchor" aria-hidden="true"></a>Processing speed</h2>
<p>Currently <em><code>epialleleR</code></em> runs in a single-thread mode only. Loading the BAM data is now done by means of <em><code>HTSlib</code></em>, therefore it is possible to speed it up using &gt;1 threads (<code>nthreads</code> option in <em><code>preprocessBam</code></em>). All operations are performed using optimised C++ functions, and usually take reasonable time. Running time for complete task “BAM on disk -&gt; CX report on disk” depends on the size of the BAM file, and the speed is usually within the range of 20-30 MB/s (or 150-200 thousand reads per second) for a single core of a relatively modern CPU (Intel(R) Core(TM) i7-7700).</p>
<p>Major bottlenecks (in BAM loading and preprocessing) were removed in the release v1.2, full multithreading and minor improvements are expected in the future.</p>
<hr>
</div>
</div>
<div id="sample-data" class="section level1">
<h1 class="hasAnchor">
<a href="#sample-data" class="anchor" aria-hidden="true"></a>Sample data</h1>
<p>The <em><code>epialleleR</code></em> package includes sample data, which was obtained using targeted sequencing. The description of assays and files is given below. All the genomic coordinates for external data files are according to GRCh38 reference assembly.</p>
<div id="amplicon-based-methylation-ngs-data" class="section level3">
<h3 class="hasAnchor">
<a href="#amplicon-based-methylation-ngs-data" class="anchor" aria-hidden="true"></a>Amplicon-based methylation NGS data</h3>
<p>The samples of Human HCT116 DKO Non-Methylated (Zymo Research, cat # D5014-1), or Human HCT116 DKO Methylated (Zymo Research, cat # D5014-2) DNA<a href="#fn4" class="footnote-ref" id="fnref4"><sup>4</sup></a>, or their mix were bisulfite-converted, and the BRCA1 gene promoter region was amplified using four pairs of primers. Amplicons were mixed, indexed and sequenced at Illumina MiSeq system. The related files are:</p>
<table class="table">
<colgroup>
<col width="33%">
<col width="33%">
<col width="33%">
</colgroup>
<thead><tr class="header">
<th>Name</th>
<th>Type</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>amplicon000meth.bam</td>
<td>BAM</td>
<td>a subset of reads for non-methylated DNA sample</td>
</tr>
<tr class="even">
<td>amplicon010meth.bam</td>
<td>BAM</td>
<td>a subset of reads for a 1:9 mix of methylated and non-methylated DNA samples</td>
</tr>
<tr class="odd">
<td>amplicon100meth.bam</td>
<td>BAM</td>
<td>a subset of reads for fully methylated DNA sample</td>
</tr>
<tr class="even">
<td>amplicon.bed</td>
<td>BED</td>
<td>genomic coordinates of four amplicons covering promoter area of BRCA1 gene</td>
</tr>
<tr class="odd">
<td>amplicon.vcf.gz</td>
<td>VCF</td>
<td>a relevant subset of sequence variations</td>
</tr>
<tr class="even">
<td>amplicon.vcf.gz.tbi</td>
<td>tabix</td>
<td>tabix file for the amplicon.vcf.gz</td>
</tr>
</tbody>
</table>
</div>
<div id="capture-based-methylation-ngs-data" class="section level3">
<h3 class="hasAnchor">
<a href="#capture-based-methylation-ngs-data" class="anchor" aria-hidden="true"></a>Capture-based methylation NGS data</h3>
<p>The tumour DNA was bisulfite-converted, fragmented and hybridized with custom-made probes covering promoter regions of 283 tumour suppressor genes (as described in <a href="#fn5" class="footnote-ref" id="fnref5"><sup>5</sup></a>). Libraries were sequenced using Illumina MiSeq system. The related files are:</p>
<table class="table">
<thead><tr class="header">
<th>Name</th>
<th>Type</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>capture.bam</td>
<td>BAM</td>
<td>a subset of reads</td>
</tr>
<tr class="even">
<td>capture.bed</td>
<td>BED</td>
<td>genomic coordinates of capture target regions</td>
</tr>
<tr class="odd">
<td>capture.vcf.gz</td>
<td>VCF</td>
<td>a relevant subset of sequence variations</td>
</tr>
<tr class="even">
<td>capture.vcf.gz.tbi</td>
<td>tabix</td>
<td>tabix file for the capture.vcf.gz</td>
</tr>
</tbody>
</table>
<hr>
</div>
</div>
<div id="typical-workflow" class="section level1">
<h1 class="hasAnchor">
<a href="#typical-workflow" class="anchor" aria-hidden="true"></a>Typical workflow</h1>
<p>As mentioned earlier, <em><code>epialleleR</code></em> uses data stored in Binary Alignment Map (BAM) files as its input. It is a prerequisite that records in the BAM file are sorted by their name (and not a genomic location), contain an XG tag with a genomic strand they map to, and an XM tag with the methylation call string — such files are produced by virtually any software tool for mapping and alignment of bisulfite sequencing reads (such as Bismark, BSMAP or Illumina DRAGEN Bio-IT Platform).</p>
<p>Please use the function help files for a full description of available parameters, as well as explanation of the function’s logic and output values.</p>
<div id="reading-the-data" class="section level2">
<h2 class="hasAnchor">
<a href="#reading-the-data" class="anchor" aria-hidden="true"></a>Reading the data</h2>
<p>All <em><code>epialleleR</code></em> methods can load BAM data using the file path. However, if a file is very large and several reports need to be prepared, it is advised to use the <em><code>preprocessBam</code></em> convenience function as shown below. This function is also used internally when a BAM file location string is supplied as an input for other <em><code>epialleleR</code></em> methods.</p>
<p><code>preprocessBam</code> currently accepts only BAM files that are derived from paired-end sequencing (create an issue if you need to process single-end BAM files). During preprocessing, paired reads are merged according to their base quality: nucleotide base with the highest value in the QUAL string is taken, unless its quality is less than <em><code>min.baseq</code></em>, which results in no information for that particular position (“-”/“N”). These <strong>merged reads</strong> are then processed as a <strong>single entity</strong> in all <em><code>epialleleR</code></em> methods. Due to merging, overlapping bases in read pairs are counted only once, and the base with the highest quality is taken.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/BBCG/epialleleR" class="external-link">epialleleR</a></span><span class="op">)</span>

<span class="va">capture.bam</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"capture.bam"</span>, package<span class="op">=</span><span class="st">"epialleleR"</span><span class="op">)</span>
<span class="va">bam.data</span>    <span class="op">&lt;-</span> <span class="fu"><a href="../reference/preprocessBam.html">preprocessBam</a></span><span class="op">(</span><span class="va">capture.bam</span><span class="op">)</span>
<span class="co">#&gt; Reading BAM file [0.027s]</span></code></pre></div>
</div>
<div id="making-cytosine-reports" class="section level2">
<h2 class="hasAnchor">
<a href="#making-cytosine-reports" class="anchor" aria-hidden="true"></a>Making cytosine reports</h2>
<p><em><code>epialleleR</code></em> can generate conventional cytosine reports in a format, which is similar to the genome-wide cytosine report produced by the <em><code>coverage2cytosine</code></em> Bismark module<a href="#fn6" class="footnote-ref" id="fnref6"><sup>6</sup></a>.</p>
<p>Please note that <em><code>generateCytosineReport</code></em> produces thresholded (VEF) report by default: <strong>methylated</strong> cytosines from reads that do <strong>not</strong> pass the threshold (<strong>hypo</strong>methylated reads) are counted as being <strong>un</strong>methylated. In order to make a conventional cytosine report, use <em><code>threshold.reads=FALSE</code></em>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># data.table::data.table object for</span>
<span class="co"># CpG VEF report</span>
<span class="va">cg.vef.report</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generateCytosineReport.html">generateCytosineReport</a></span><span class="op">(</span><span class="va">bam.data</span><span class="op">)</span>
<span class="co">#&gt; Already preprocessed BAM supplied as an input. Options 'min.mapq', 'min.baseq', 'skip.duplicates' and 'nthreads' will have no effect.</span>
<span class="co">#&gt; Thresholding reads [0.002s]</span>
<span class="co">#&gt; Preparing cytosine report [0.036s]</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">cg.vef.report</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">meth</span><span class="op">+</span><span class="va">unmeth</span>, decreasing<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>
<span class="co">#&gt;    rname strand      pos context meth unmeth</span>
<span class="co">#&gt; 1: chr17      + 61864475      CG    7      9</span>
<span class="co">#&gt; 2: chr17      + 61864486      CG   10      6</span>
<span class="co">#&gt; 3: chr17      + 61864504      CG    9      7</span>
<span class="co">#&gt; 4: chr20      - 57267455      CG   13      1</span>
<span class="co">#&gt; 5: chr17      - 61863826      CG    0     13</span>
<span class="co">#&gt; 6: chr17      - 61863830      CG    0     13</span>

<span class="co"># CpG cytosine report</span>
<span class="va">cg.report</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generateCytosineReport.html">generateCytosineReport</a></span><span class="op">(</span><span class="va">bam.data</span>, threshold.reads<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>
<span class="co">#&gt; Already preprocessed BAM supplied as an input. Options 'min.mapq', 'min.baseq', 'skip.duplicates' and 'nthreads' will have no effect.</span>
<span class="co">#&gt; Preparing cytosine report [0.029s]</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">cg.report</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">meth</span><span class="op">+</span><span class="va">unmeth</span>, decreasing<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>
<span class="co">#&gt;    rname strand      pos context meth unmeth</span>
<span class="co">#&gt; 1: chr17      + 61864475      CG    8      8</span>
<span class="co">#&gt; 2: chr17      + 61864486      CG   10      6</span>
<span class="co">#&gt; 3: chr17      + 61864504      CG   10      6</span>
<span class="co">#&gt; 4: chr20      - 57267455      CG   13      1</span>
<span class="co">#&gt; 5: chr17      - 61863826      CG    0     13</span>
<span class="co">#&gt; 6: chr17      - 61863830      CG    0     13</span>

<span class="co"># CX cytosine report</span>
<span class="va">cx.report</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generateCytosineReport.html">generateCytosineReport</a></span><span class="op">(</span><span class="va">bam.data</span>, threshold.reads<span class="op">=</span><span class="cn">FALSE</span>,
                                    report.context<span class="op">=</span><span class="st">"CX"</span><span class="op">)</span>
<span class="co">#&gt; Already preprocessed BAM supplied as an input. Options 'min.mapq', 'min.baseq', 'skip.duplicates' and 'nthreads' will have no effect.</span>
<span class="co">#&gt; Preparing cytosine report [0.042s]</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">cx.report</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">meth</span><span class="op">+</span><span class="va">unmeth</span>, decreasing<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>
<span class="co">#&gt;    rname strand      pos context meth unmeth</span>
<span class="co">#&gt; 1: chr17      + 61864338     CHG    1     25</span>
<span class="co">#&gt; 2: chr17      + 61864348     CHH    0     24</span>
<span class="co">#&gt; 3: chr17      + 61864364     CHH    0     24</span>
<span class="co">#&gt; 4: chr17      + 61864365     CHH    0     24</span>
<span class="co">#&gt; 5: chr17      + 61864373     CHH    0     24</span>
<span class="co">#&gt; 6: chr17      + 61864324     CHG    0     23</span></code></pre></div>
</div>
<div id="making-vef-reports-for-a-set-of-genomic-regions" class="section level2">
<h2 class="hasAnchor">
<a href="#making-vef-reports-for-a-set-of-genomic-regions" class="anchor" aria-hidden="true"></a>Making VEF reports for a set of genomic regions</h2>
<p><em><code>epialleleR</code></em> allows to make reports not only for individual cytosine bases, but also for a set of genomic regions. It is especially useful when the targeted methylation sequencing was used to produce reads (such as amplicon sequencing or hybridization capture using, e.g., Agilent SureSelect Target Enrichment Probes).</p>
<p>The amplicon sequencing principally differs from capture-based assays in that the coordinates of reads are known. Therefore, reads can be assigned to amplicons by their exact positions, while to the capture targets — by the overlap. For this, <em><code>epialleleR</code></em> provides generic <em><code>generateBedReport</code></em> function as well as two of its aliases, <em><code>generateAmpliconReport</code></em> (for amplicon-based NGS) and <em><code>generateCaptureReport</code></em> (for capture-based NGS).</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># report for amplicon-based data</span>
<span class="co"># matching is done by exact start or end positions plus/minus tolerance</span>
<span class="va">amplicon.report</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generateBedReport.html">generateAmpliconReport</a></span><span class="op">(</span>
  bam<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"amplicon010meth.bam"</span>, package<span class="op">=</span><span class="st">"epialleleR"</span><span class="op">)</span>,
  bed<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"amplicon.bed"</span>, package<span class="op">=</span><span class="st">"epialleleR"</span><span class="op">)</span>
<span class="op">)</span>
<span class="co">#&gt; Reading BED file [0.064s]</span>
<span class="co">#&gt; Reading BAM file [0.007s]</span>
<span class="co">#&gt; Thresholding reads [0.001s]</span>
<span class="co">#&gt; Preparing amplicon report [0.057s]</span>
<span class="va">amplicon.report</span>
<span class="co">#&gt;    seqnames    start      end width strand amplicon nreads+ nreads-        VEF</span>
<span class="co">#&gt; 1:    chr17 43125624 43126026   403      * CpG00-13       0     156 0.08333333</span>
<span class="co">#&gt; 2:    chr17 43125270 43125640   371      * CpG14-31       0      61 0.11475410</span>
<span class="co">#&gt; 3:    chr17 43125171 43125550   380      * CpG17-34       0      93 0.05376344</span>
<span class="co">#&gt; 4:    chr17 43124861 43125249   389      * CpG33-49       0      84 0.10714286</span>
<span class="co">#&gt; 5:     &lt;NA&gt;       NA       NA    NA   &lt;NA&gt;     &lt;NA&gt;      60      46 0.13207547</span>

<span class="co"># report for capture-based data</span>
<span class="co"># matching is done by overlap</span>
<span class="va">capture.report</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generateBedReport.html">generateCaptureReport</a></span><span class="op">(</span>
  bam<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"capture.bam"</span>, package<span class="op">=</span><span class="st">"epialleleR"</span><span class="op">)</span>,
  bed<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"capture.bed"</span>, package<span class="op">=</span><span class="st">"epialleleR"</span><span class="op">)</span>
<span class="op">)</span>
<span class="co">#&gt; Reading BED file [0.043s]</span>
<span class="co">#&gt; Reading BAM file [0.024s]</span>
<span class="co">#&gt; Thresholding reads [0.001s]</span>
<span class="co">#&gt; Preparing capture report [0.026s]</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">capture.report</span><span class="op">)</span>
<span class="co">#&gt;    seqnames    start      end width strand     V4 nreads+ nreads-       VEF</span>
<span class="co">#&gt; 1:     chr1  3067647  3069703  2057      * PRDM16       2       1 1.0000000</span>
<span class="co">#&gt; 2:     chr1  3651039  3653096  2058      *   TP73       0       2 0.5000000</span>
<span class="co">#&gt; 3:     chr1  3689153  3691202  2050      *   TP73       0       2 1.0000000</span>
<span class="co">#&gt; 4:     chr1  3696519  3698570  2052      *   TP73       1       2 1.0000000</span>
<span class="co">#&gt; 5:     chr1  6179609  6181670  2062      *   CHD5       0       3 0.6666667</span>
<span class="co">#&gt; 6:     chr1 13698869 13699064   196      *  PRDM2      NA      NA        NA</span>

<span class="co"># generateBedReport is a generic function for BED-guided reports</span>
<span class="va">bed.report</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generateBedReport.html">generateBedReport</a></span><span class="op">(</span>
  bam<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"capture.bam"</span>, package<span class="op">=</span><span class="st">"epialleleR"</span><span class="op">)</span>,
  bed<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"capture.bed"</span>, package<span class="op">=</span><span class="st">"epialleleR"</span><span class="op">)</span>,
  bed.type<span class="op">=</span><span class="st">"capture"</span>
<span class="op">)</span>
<span class="co">#&gt; Reading BED file [0.017s]</span>
<span class="co">#&gt; Reading BAM file [0.024s]</span>
<span class="co">#&gt; Thresholding reads [0.001s]</span>
<span class="co">#&gt; Preparing capture report [0.025s]</span>
<span class="fu"><a href="https://rdrr.io/r/base/identical.html" class="external-link">identical</a></span><span class="op">(</span><span class="va">capture.report</span>, <span class="va">bed.report</span><span class="op">)</span>
<span class="co">#&gt; [1] TRUE</span></code></pre></div>
</div>
<div id="plotting-the-distribution-of-per-read-beta-values" class="section level2">
<h2 class="hasAnchor">
<a href="#plotting-the-distribution-of-per-read-beta-values" class="anchor" aria-hidden="true"></a>Plotting the distribution of per-read beta values</h2>
<p>As stated in the introduction, human genomic DNA regions often show bimodal methylation patterns. <em><code>epialleleR</code></em> allows to visualize this information by plotting empirical cumulative distribution functions (eCDFs) for within- and out-of-context beta values.</p>
<p>The code below produces plots for the sequencing reads of control DNA samples. Note that within-the-context eCDF(0.5) values are very close to the expected 1-VEF values for the corresponding control DNA samples:</p>
<ul>
<li>non-methylated DNA — expected VEF = 0, observed 1-eCDF(0.5) ≈ 0</li>
<li>1:9 mix of methylated and non-methylated DNA — expected VEF = 0.1, observed 1-eCDF(0.5) ≈ 0.1</li>
<li>and fully methylated DNA — expected VEF = 1, observed 1-eCDF(0.5) ≈ 1</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># First, let's visualise eCDFs for within- and out-of-context beta values</span>
<span class="co"># for all four amplicons and unmatched reads. Note that within-the-context eCDF</span>
<span class="co"># of 0.5 is very close to the expected 1-VEF value (0.1) for all amplicons</span>
<span class="co"># produced from this 1:9 mix of methylated and non-methylated control DNA</span>

<span class="co"># let's compute eCDF</span>
<span class="va">amplicon.ecdfs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generateBedEcdf.html">generateBedEcdf</a></span><span class="op">(</span>
  bam<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"amplicon010meth.bam"</span>, package<span class="op">=</span><span class="st">"epialleleR"</span><span class="op">)</span>,
  bed<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"amplicon.bed"</span>, package<span class="op">=</span><span class="st">"epialleleR"</span><span class="op">)</span>,
  bed.rows<span class="op">=</span><span class="cn">NULL</span>
<span class="op">)</span>
<span class="co">#&gt; Reading BED file [0.015s]</span>
<span class="co">#&gt; Reading BAM file [0.007s]</span>
<span class="co">#&gt; Computing ECDFs for within- and out-of-context per-read beta values [0.018s]</span>

<span class="co"># there are 5 items in amplicon.ecdfs, let's plot all of them</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">amplicon.ecdfs</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>

<span class="co"># cycle through items</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">x</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">amplicon.ecdfs</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
  <span class="co"># four of them have names corresponding to genomic regions of amplicon.bed</span>
  <span class="co"># fifth - NA for all the reads that don't match to any of those regions</span>
  <span class="va">main</span> <span class="op">&lt;-</span> <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">amplicon.ecdfs</span><span class="op">[</span><span class="va">x</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="st">"unmatched"</span>
          <span class="kw">else</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">amplicon.ecdfs</span><span class="op">[</span><span class="va">x</span><span class="op">]</span><span class="op">)</span>
  
  <span class="co"># plotting eCDF for within-the-context per-read beta values (in red)</span>
  <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">amplicon.ecdfs</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">context</span>, col<span class="op">=</span><span class="st">"red"</span>, verticals<span class="op">=</span><span class="cn">TRUE</span>, do.points<span class="op">=</span><span class="cn">FALSE</span>,
       xlim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>, xlab<span class="op">=</span><span class="st">"per-read beta value"</span>, ylab<span class="op">=</span><span class="st">"cumulative density"</span>,
       main<span class="op">=</span><span class="va">main</span><span class="op">)</span>
  
  <span class="co"># adding eCDF for out-of-context per-read beta values (in blue)</span>
  <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">amplicon.ecdfs</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">out.of.context</span>, add<span class="op">=</span><span class="cn">TRUE</span>, col<span class="op">=</span><span class="st">"blue"</span>,
       verticals<span class="op">=</span><span class="cn">TRUE</span>, do.points<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<p><img src="epialleleR_files/figure-html/unnamed-chunk-5-1.png" width="100%" height="100%"></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R">

<span class="co"># Second, let's compare eCDFs for within-the-context beta values for only one</span>
<span class="co"># amplicon but all three sequenced samples: pure non-methylated DNA, 1:9 mix of</span>
<span class="co"># methylated and non-methylated DNA, and fully methylated DNA</span>

<span class="co"># our files</span>
<span class="va">bam.files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"amplicon000meth.bam"</span>, <span class="st">"amplicon010meth.bam"</span>,
               <span class="st">"amplicon100meth.bam"</span><span class="op">)</span>

<span class="co"># let's plot all of them</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">bam.files</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>

<span class="co"># cycle through items</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">f</span> <span class="kw">in</span> <span class="va">bam.files</span><span class="op">)</span> <span class="op">{</span>
  <span class="co"># let's compute eCDF</span>
  <span class="va">amplicon.ecdfs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generateBedEcdf.html">generateBedEcdf</a></span><span class="op">(</span>
    bam<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="va">f</span>, package<span class="op">=</span><span class="st">"epialleleR"</span><span class="op">)</span>,
    bed<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"amplicon.bed"</span>, package<span class="op">=</span><span class="st">"epialleleR"</span><span class="op">)</span>,
    <span class="co"># only the second amplicon</span>
    bed.rows<span class="op">=</span><span class="fl">2</span>, verbose<span class="op">=</span><span class="cn">FALSE</span>
  <span class="op">)</span>
  
  <span class="co"># plotting eCDF for within-the-context per-read beta values (in red)</span>
  <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">amplicon.ecdfs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">context</span>, col<span class="op">=</span><span class="st">"red"</span>, verticals<span class="op">=</span><span class="cn">TRUE</span>, do.points<span class="op">=</span><span class="cn">FALSE</span>,
       xlim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>, xlab<span class="op">=</span><span class="st">"per-read beta value"</span>, ylab<span class="op">=</span><span class="st">"cumulative density"</span>,
       main<span class="op">=</span><span class="va">f</span><span class="op">)</span>
  
   <span class="co"># adding eCDF for out-of-context per-read beta values (in blue)</span>
  <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">amplicon.ecdfs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">out.of.context</span>, add<span class="op">=</span><span class="cn">TRUE</span>, col<span class="op">=</span><span class="st">"blue"</span>,
       verticals<span class="op">=</span><span class="cn">TRUE</span>, do.points<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<p><img src="epialleleR_files/figure-html/unnamed-chunk-5-2.png" width="100%" height="100%"></p>
</div>
<div id="exploring-sequence-variants-in-epialleles" class="section level2">
<h2 class="hasAnchor">
<a href="#exploring-sequence-variants-in-epialleles" class="anchor" aria-hidden="true"></a>Exploring sequence variants in epialleles</h2>
<p>It is known that sequence variants can affect the methylation status of a DNA<a href="#fn7" class="footnote-ref" id="fnref7"><sup>7</sup></a>. The <em><code>generateVcfReport</code></em> function calculates frequencies of single nucleotide variants (SNVs) within epialleles and tests for the association between SNV and epiallelic status using Fisher Exact test. Base counts and the test’s p-values are included in the returned value.</p>
<p>In addition to BAM file location string or preprocessed BAM object, the function requires a location string for the Variant Call Format (VCF) file or the VCF object that was obtained using <em><code><a href="https://rdrr.io/pkg/VariantAnnotation/man/readVcf-methods.html" class="external-link">VariantAnnotation::readVcf</a></code></em> function. As VCF files can be extremely large, it is strongly advised to prefilter the VCF object by the relevant set of genomic regions, or specify such relevant set of regions as a <em><code>bed</code></em> parameter when <em><code>vcf</code></em> points to a VCF file location.</p>
<p>Please note, that the output report is currently limited to SNVs only. Also, the default (<code>min.baseq=0</code>) output of <code>generateVcfReport</code> is equivalent to the one of <code>samtools mplieup -Q 0 ...</code>, and therefore may result in false SNVs caused by misalignments. Remember to increase <code>min.baseq</code> (<code>samtools mplieup -Q</code> default value is 13) to obtain results of a higher quality.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># VCF report</span>
<span class="va">vcf.report</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generateVcfReport.html">generateVcfReport</a></span><span class="op">(</span>
  bam<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"amplicon010meth.bam"</span>, package<span class="op">=</span><span class="st">"epialleleR"</span><span class="op">)</span>,
  bed<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"amplicon.bed"</span>, package<span class="op">=</span><span class="st">"epialleleR"</span><span class="op">)</span>,
  vcf<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"amplicon.vcf.gz"</span>, package<span class="op">=</span><span class="st">"epialleleR"</span><span class="op">)</span>,
  <span class="co"># thresholds on alignment and base quality</span>
  min.mapq<span class="op">=</span><span class="fl">30</span>, min.baseq<span class="op">=</span><span class="fl">13</span>,
  <span class="co"># when VCF seqlevels are different from BED and BAM it is possible</span>
  <span class="co"># to convert them internally</span>
  vcf.style<span class="op">=</span><span class="st">"NCBI"</span>
<span class="op">)</span>
<span class="co">#&gt; Reading BED file [0.015s]</span>
<span class="co">#&gt; Reading VCF file [1.016s]</span>
<span class="co">#&gt; Reading BAM file [0.032s]</span>
<span class="co">#&gt; Thresholding reads [0.001s]</span>
<span class="co">#&gt; Extracting base frequences [0.055s]</span>

<span class="co"># NA values are shown for the C-&gt;T variants on the "+" and G-&gt;A on the "-"</span>
<span class="co"># strands, because bisulfite conversion makes their counting impossible</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">vcf.report</span><span class="op">)</span>
<span class="co">#&gt;           name seqnames    range REF ALT M+Ref U+Ref M-Ref U-Ref M+Alt U+Alt M-Alt U-Alt SumRef</span>
<span class="co">#&gt; 1: rs546660277    chr17 43124874   A   C     0     0     9    74     0     0     0     0     83</span>
<span class="co">#&gt; 2: rs574263814    chr17 43124891   G   A     0     0    NA    NA     0     0    NA    NA      0</span>
<span class="co">#&gt; 3:   rs8176076    chr17 43124935   G   A     0     0    NA    NA     0     0    NA    NA      0</span>
<span class="co">#&gt; 4: rs535977743    chr17 43125016   C   T    NA    NA     9    73    NA    NA     0     0     82</span>
<span class="co">#&gt; 5: rs191784032    chr17 43125050   C   A     0     0     9    73     0     0     0     1     82</span>
<span class="co">#&gt; 6: rs111956204    chr17 43125083   C   A     0     0     9    74     0     0     0     0     83</span>
<span class="co">#&gt;    SumAlt FEp+ FEp-</span>
<span class="co">#&gt; 1:      0    1    1</span>
<span class="co">#&gt; 2:      0    1   NA</span>
<span class="co">#&gt; 3:      0    1   NA</span>
<span class="co">#&gt; 4:      0   NA    1</span>
<span class="co">#&gt; 5:      1    1    1</span>
<span class="co">#&gt; 6:      0    1    1</span>

<span class="co"># let's sort the report by increasing Fisher's exact test's p-values.</span>
<span class="co"># the p-values are given separately for reads that map to the "+"</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">vcf.report</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">`FEp-`</span>, na.last<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>
<span class="co">#&gt;           name seqnames    range REF ALT M+Ref U+Ref M-Ref U-Ref M+Alt U+Alt M-Alt U-Alt SumRef</span>
<span class="co">#&gt; 1: rs546660277    chr17 43124874   A   C     0     0     9    74     0     0     0     0     83</span>
<span class="co">#&gt; 2: rs535977743    chr17 43125016   C   T    NA    NA     9    73    NA    NA     0     0     82</span>
<span class="co">#&gt; 3: rs191784032    chr17 43125050   C   A     0     0     9    73     0     0     0     1     82</span>
<span class="co">#&gt; 4: rs111956204    chr17 43125083   C   A     0     0     9    74     0     0     0     0     83</span>
<span class="co">#&gt; 5:  rs55680227    chr17 43125086   A   C     0     0     8    64     0     0     0     0     72</span>
<span class="co">#&gt; 6: rs539733232    chr17 43125088   C   A     0     0     8    71     0     0     0     0     79</span>
<span class="co">#&gt;    SumAlt FEp+ FEp-</span>
<span class="co">#&gt; 1:      0    1    1</span>
<span class="co">#&gt; 2:      0   NA    1</span>
<span class="co">#&gt; 3:      1    1    1</span>
<span class="co">#&gt; 4:      0    1    1</span>
<span class="co">#&gt; 5:      0    1    1</span>
<span class="co">#&gt; 6:      0    1    1</span>

<span class="co"># and to the "-" strand</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">vcf.report</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">`FEp+`</span>, na.last<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>
<span class="co">#&gt;           name seqnames    range REF ALT M+Ref U+Ref M-Ref U-Ref M+Alt U+Alt M-Alt U-Alt SumRef</span>
<span class="co">#&gt; 1: rs546660277    chr17 43124874   A   C     0     0     9    74     0     0     0     0     83</span>
<span class="co">#&gt; 2: rs574263814    chr17 43124891   G   A     0     0    NA    NA     0     0    NA    NA      0</span>
<span class="co">#&gt; 3:   rs8176076    chr17 43124935   G   A     0     0    NA    NA     0     0    NA    NA      0</span>
<span class="co">#&gt; 4: rs191784032    chr17 43125050   C   A     0     0     9    73     0     0     0     1     82</span>
<span class="co">#&gt; 5: rs111956204    chr17 43125083   C   A     0     0     9    74     0     0     0     0     83</span>
<span class="co">#&gt; 6:  rs55680227    chr17 43125086   A   C     0     0     8    64     0     0     0     0     72</span>
<span class="co">#&gt;    SumAlt FEp+ FEp-</span>
<span class="co">#&gt; 1:      0    1    1</span>
<span class="co">#&gt; 2:      0    1   NA</span>
<span class="co">#&gt; 3:      0    1   NA</span>
<span class="co">#&gt; 4:      1    1    1</span>
<span class="co">#&gt; 5:      0    1    1</span>
<span class="co">#&gt; 6:      0    1    1</span></code></pre></div>
<hr>
</div>
</div>
<div id="other-information" class="section level1">
<h1 class="hasAnchor">
<a href="#other-information" class="anchor" aria-hidden="true"></a>Other information</h1>
<div id="citing-the-epialleler-package" class="section level2">
<h2 class="hasAnchor">
<a href="#citing-the-epialleler-package" class="anchor" aria-hidden="true"></a>Citing the <em><code>epialleleR</code></em> package</h2>
<p>The experimental data analysed using the package has not been published yet. The citation information will be updated in the nearest future.</p>
</div>
<div id="session-info" class="section level2">
<h2 class="hasAnchor">
<a href="#session-info" class="anchor" aria-hidden="true"></a>Session Info</h2>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R version 4.1.1 (2021-08-10)</span>
<span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">#&gt; Running under: Ubuntu 20.04.3 LTS</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       </span>
<span class="co">#&gt;  [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             </span>
<span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              </span>
<span class="co">#&gt; [10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt; [1] GenomeInfoDb_1.29.8 IRanges_2.27.2      S4Vectors_0.31.5    BiocGenerics_0.39.2</span>
<span class="co">#&gt; [5] epialleleR_1.1.11  </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;  [1] bitops_1.0-7                matrixStats_0.61.0          fs_1.5.0                   </span>
<span class="co">#&gt;  [4] bit64_4.0.5                 filelock_1.0.2              progress_1.2.2             </span>
<span class="co">#&gt;  [7] httr_1.4.2                  rprojroot_2.0.2             tools_4.1.1                </span>
<span class="co">#&gt; [10] bslib_0.3.0                 utf8_1.2.2                  R6_2.5.1                   </span>
<span class="co">#&gt; [13] DBI_1.1.1                   tidyselect_1.1.1            prettyunits_1.1.1          </span>
<span class="co">#&gt; [16] bit_4.0.4                   curl_4.3.2                  compiler_4.1.1             </span>
<span class="co">#&gt; [19] textshaping_0.3.5           Biobase_2.53.0              xml2_1.3.2                 </span>
<span class="co">#&gt; [22] desc_1.4.0                  DelayedArray_0.19.4         rtracklayer_1.53.1         </span>
<span class="co">#&gt; [25] sass_0.4.0                  rappdirs_0.3.3              pkgdown_1.9000.9000.9000   </span>
<span class="co">#&gt; [28] systemfonts_1.0.2           stringr_1.4.0               digest_0.6.28              </span>
<span class="co">#&gt; [31] Rsamtools_2.9.1             rmarkdown_2.11              XVector_0.33.0             </span>
<span class="co">#&gt; [34] pkgconfig_2.0.3             htmltools_0.5.2             MatrixGenerics_1.5.4       </span>
<span class="co">#&gt; [37] highr_0.9                   dbplyr_2.1.1                fastmap_1.1.0              </span>
<span class="co">#&gt; [40] BSgenome_1.61.0             rlang_0.4.11                RSQLite_2.2.8              </span>
<span class="co">#&gt; [43] jquerylib_0.1.4             BiocIO_1.3.0                generics_0.1.0             </span>
<span class="co">#&gt; [46] jsonlite_1.7.2              BiocParallel_1.27.12        dplyr_1.0.7                </span>
<span class="co">#&gt; [49] VariantAnnotation_1.39.0    RCurl_1.98-1.5              magrittr_2.0.1             </span>
<span class="co">#&gt; [52] GenomeInfoDbData_1.2.7      Matrix_1.3-4                Rcpp_1.0.7                 </span>
<span class="co">#&gt; [55] fansi_0.5.0                 lifecycle_1.0.1             stringi_1.7.4              </span>
<span class="co">#&gt; [58] yaml_2.2.1                  SummarizedExperiment_1.23.4 zlibbioc_1.39.0            </span>
<span class="co">#&gt; [61] BiocFileCache_2.1.1         grid_4.1.1                  blob_1.2.2                 </span>
<span class="co">#&gt; [64] parallel_4.1.1              crayon_1.4.1                lattice_0.20-45            </span>
<span class="co">#&gt; [67] Biostrings_2.61.2           GenomicFeatures_1.45.2      hms_1.1.1                  </span>
<span class="co">#&gt; [70] KEGGREST_1.33.0             knitr_1.36                  pillar_1.6.3               </span>
<span class="co">#&gt; [73] GenomicRanges_1.45.0        rjson_0.2.20                biomaRt_2.49.4             </span>
<span class="co">#&gt; [76] XML_3.99-0.8                glue_1.4.2                  evaluate_0.14              </span>
<span class="co">#&gt; [79] data.table_1.14.2           png_0.1-7                   vctrs_0.3.8                </span>
<span class="co">#&gt; [82] purrr_0.3.4                 assertthat_0.2.1            cachem_1.0.6               </span>
<span class="co">#&gt; [85] xfun_0.26                   restfulr_0.0.13             ragg_1.1.3                 </span>
<span class="co">#&gt; [88] tibble_3.1.5                GenomicAlignments_1.29.0    AnnotationDbi_1.55.1       </span>
<span class="co">#&gt; [91] memoise_2.0.0               ellipsis_0.3.2</span></code></pre></div>
<hr>
</div>
</div>
<div id="references" class="section level1">
<h1 class="hasAnchor">
<a href="#references" class="anchor" aria-hidden="true"></a>References</h1>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p><a href="https://doi.org/10.1146/annurev.pharmtox.45.120403.095832" class="external-link uri">https://doi.org/10.1146/annurev.pharmtox.45.120403.095832</a><a href="#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p><a href="https://doi.org/10.1101/2020.12.01.403501" class="external-link uri">https://doi.org/10.1101/2020.12.01.403501</a><a href="#fnref2" class="footnote-back">↩︎</a></p></li>
<li id="fn3"><p><a href="https://doi.org/10.1093/bib/bbx077" class="external-link uri">https://doi.org/10.1093/bib/bbx077</a><a href="#fnref3" class="footnote-back">↩︎</a></p></li>
<li id="fn4"><p><a href="https://www.zymoresearch.com/products/human-methylated-non-methylated-dna-set-dna-w-primers" class="external-link uri">https://www.zymoresearch.com/products/human-methylated-non-methylated-dna-set-dna-w-primers</a><a href="#fnref4" class="footnote-back">↩︎</a></p></li>
<li id="fn5"><p><a href="https://dx.doi.org/10.1186%2Fs13148-020-00920-7" class="external-link uri">https://dx.doi.org/10.1186%2Fs13148-020-00920-7</a><a href="#fnref5" class="footnote-back">↩︎</a></p></li>
<li id="fn6"><p><a href="https://www.bioinformatics.babraham.ac.uk/projects/bismark/" class="external-link uri">https://www.bioinformatics.babraham.ac.uk/projects/bismark/</a><a href="#fnref6" class="footnote-back">↩︎</a></p></li>
<li id="fn7"><p><a href="https://doi.org/10.1038/modpathol.2009.130" class="external-link uri">https://doi.org/10.1038/modpathol.2009.130</a><a href="#fnref7" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Oleksii Nikolaienko.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link external-link">pkgdown</a> 1.9000.9000.9000.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
