#' generateCytosineReport
#'
#' @description
#' `generateCytosineReport` This function counts methylated and unmethylated
#' DNA bases taking into the account average methylation level of the entire
#' sequence read
#'
#' @details
#' The function reports cytosine methylation information using BAM file as an
#' input. In contrast to the other currently available software, reads can be
#' thresholded by their average methylation level before counting methylated
#' bases, effectively resulting in hypermethylated Variant Epiallele Frequency
#' (VEF) being reported instead of beta value. Here's how it works:
#' Let's suppose we have a BAM file with four reads mapped to chromosome 1 
#' positions 1-16. Assuming the default thresholding parameters (threshold.reads
#' = TRUE, threshold.context = "CG", min.context.sites = 2, min.context.beta =
#' 0.5, max.outofcontext.beta = 0.1), the input and results will look as
#' following:
#' 
#' | read chr:strand:start | methylation string | threshold | reason | methylation reported |
#' | chr1:+:1 | ...Z..x+.h..x..h. | below | min.context.sites < 2 (only one "zZ" base) | C at position 4: unmethylated |
#' | chr1:+:1 | ...Z..z.h..x..h.  | above | pass all criteria | C4: methylated, C7: unmethylated |
#' | chr1:+:1 | ...Z..z.h..X..h.  | below | max.outofcontext.beta > 0.1 (1XH / 3xXhH = 0.33) | C4 and C7: unmethylated |
#' | chr1:+:1 | ...Z..z.h..z-.h.  | below | min.context.beta < 0.5 (1Z / 3zZ = 0.33) | C4 and C7: unmethylated |
#' 
#' Only the second read will satisfy all the thresholding criteria, leading to
#' the following CX report:
#' 
#' | rname | strand | pos | context | meth | unmeth | triad |
#' | chr1 | + | 4 | CG | 1 | 3 | NNN |
#' | chr1 | + | 7 | CG | 0 | 3 | NNN |
#' | chr1 | + | 9 | CHH | 0 | 4 | NNN |
#' | chr1 | + | 12 | CHG | 0 | 3 | NNN |
#' | chr1 | + | 15 | CHH | 0 | 4 | NNN |
#' 
#' With the thresholding disabled (threshold.reads = FALSE) all methylated bases
#' will retain their status, so the CX report will be similar to the reports
#' produced by other methylation callers (such as Bismark or Illumina DRAGEN Bio
#' IT Platform):
#' 
#' | rname | strand | pos | context | meth | unmeth | triad |
#' | chr1 | + | 4 | CG | 4 | 0 | NNN |
#' | chr1 | + | 7 | CG | 0 | 3 | NNN |
#' | chr1 | + | 9 | CHH | 0 | 4 | NNN |
#' | chr1 | + | 12 | CHG | 1 | 2 | NNN |
#' | chr1 | + | 15 | CHH | 0 | 4 | NNN |
#' 
#' Other notes:
#' 
#' Methylation string bases in unknown context ("uU") are simply ignored, which,
#' to the best of our knowledge, is consistent with the behaviour of other
#' tools.
#' 
#' In order to mitigate sequencing errors (leading to rare variations in the
#' methylation context, as in reads 1 and 4 above), the context present in more
#' than 50% of the reads is assumed to be correct, while all bases at the same
#' position but having other methylation context are simply ignored. This allows
#' reports to be prepared without using the reference genome sequence.
#' 
#' The downside of not using the reference genome sequence is the inability to
#' determine the actual sequnce of triplet for every base in the cytosine 
#' report. For now the sequence is reported as "NNN" and this will stay until
#' such information will be considered as worth adding. 
#' 
#' Please also note that for all its methods, `epialleleR` requires methylation
#' call string to be present in a BAM file - i.e. methylation calling must be
#' performed after read mapping/alignment by your software of choice.
#'
#' @param bam BAM file location string OR preprocessed output of
#' \code{\link{preprocessBam}} function
#' @param report.file file location string to write the cytosine report. If NULL
#' (the default) then report is returned as a
#' \code{\link[data.table]{data.table}} object
#' @param threshold.reads boolean defining if sequence reads should be
#' thresholded before counting methylated cytosines (default: TRUE). Disabling
#' thresholding makes the report virtually indistinguishable from the ones
#' generated by other software, such as Bismark or Illumina DRAGEN Bio IT
#' Platform
#' @param threshold.context string defining cytosine methylation context used
#' for thresholdning the reads.
#' \itemize{
#'   \item "CG" (the default) -- within-the-context: CpG cytosines (called as
#'   zZ), out-of-context: all the other cytosines (hHxX)
#'   \item "CHG" -- within-the-context: CHG cytosines (xX), out-of-context: hHzZ
#'   \item "CHH" -- within-the-context: CHH cytosines (hH), out-of-context: xXzZ
#'   \item "CxG" -- within-the-context: CG and CHG cytosines (zZxX),
#'   out-of-context: CHH cytosines (hH)
#'   \item "CX" -- all cytosines are considered within-the-context, this
#'   effectively results in no thresholdning
#' }
#' This option has no effect when read thresholding is disabled
#' @param min.context.sites non-negative integer for minimum number of cytosines
#' within the `threshold.context` (default: 2). Reads containing *fewer*
#' within-the-context cytosines are considered completely unmethylated (all C
#' are counted as T). This option has no effect when read thresholding is
#' disabled
#' @param min.context.beta real number in the range [0;1] (default: 0.5). Reads
#' with average beta value for within-the-context cytosines *below* this
#' threshold are considered completely unmethylated (all C are counted as T).
#' This option has no effect when read thresholding is disabled
#' @param max.outofcontext.beta real number in the range [0;1] (default: 0.1).
#' Reads with average beta value for out-of-context cytosines *above* this
#' threshold are considered completely unmethylated (all C are counted as T).
#' This option has no effect when read thresholding is disabled
#' @param report.context string defining cytosine methylation context to report
#' (default: value of `threshold.context`)
#' @param min.mapq non-negative integer threshold for minimum read mapping
#' quality (default: 0). Option has no effect if preprocessed BAM data was
#' supplied as an input
#' @param skip.duplicates boolean defining if duplicate aligned reads should be
#' skipped (default: FALSE). Option has no effect if preprocessed BAM data was
#' supplied as an input OR duplicate reads were not marked by alignment software
#' @param gzip boolean to compress the report (default: FALSE)
#' @param verbose boolean to report progress and timings (default: TRUE)
#' @return \code{\link[data.table]{data.table}} object containing cytosine
#' report in Bismark format or NULL if report.file was specified
#' @seealso \code{\link{preprocessBam}}
#' @examples
#'   capture.bam <- system.file("extdata", "capture.bam", package="epialleleR")
#'   cg.report   <- generateCytosineReport(capture.bam)
#'   cx.report   <- generateCytosineReport(capture.bam, threshold.reads=FALSE,
#'                  report.context="CX")
#' @export
generateCytosineReport <- function (bam,
                                    report.file=NULL,
                                    threshold.reads=TRUE,
                                    threshold.context=c("CG", "CHG", "CHH", "CxG", "CX"),
                                    min.context.sites=2,
                                    min.context.beta=0.5,
                                    max.outofcontext.beta=0.1, # double 0 to 1
                                    report.context=threshold.context,
                                    min.mapq=0,
                                    skip.duplicates=FALSE,
                                    gzip=FALSE,
                                    verbose=TRUE)
{
  threshold.context <- match.arg(threshold.context, threshold.context)
  report.context    <- match.arg(report.context, report.context)
  
  bam <- preprocessBam(bam.file=bam, min.mapq=min.mapq,
                       skip.duplicates=skip.duplicates, verbose=verbose)
  
  if (threshold.reads) {
    bam$pass <- .thresholdReads(
      bam.processed=bam,
      ctx.meth=.context.to.bases[[threshold.context]][["ctx.meth"]],
      ctx.unmeth=.context.to.bases[[threshold.context]][["ctx.unmeth"]],
      ooctx.meth=.context.to.bases[[threshold.context]][["ooctx.meth"]],
      ooctx.unmeth=.context.to.bases[[threshold.context]][["ooctx.unmeth"]],
      min.context.sites=min.context.sites,
      min.context.beta=min.context.beta,
      max.outofcontext.beta=max.outofcontext.beta,
      verbose=verbose
    )
  } else {
    bam$pass <- TRUE
  }
  
  cx.report <- .getCytosineReport(
    bam.processed=bam,
    ctx=paste0(.context.to.bases[[report.context]][["ctx.meth"]],
               .context.to.bases[[report.context]][["ctx.unmeth"]]),
    verbose=verbose
  )
  
  # clean up by genome context below!
  # if genome not avail, remove the least frequent context?
  
  if (is.null(report.file))
    return(cx.report)
  else
    .writeReport(report=cx.report, report.file=report.file, gzip=gzip,
                 verbose=verbose)
}


#|[c]{}^

# ##############################################################################
